---
title: "04-TE-meth-intersect"
output: html_document
---

test run on one

```{bash}
pwd

```

```{bash}
/home/shared/bedtools2/bin/intersectBed \
-wb \
-a ../data/3F_R1_val_1_10x.bedgraph \
-b ../genome-features/C_virginica-3.0_TE-all.gff \
> ../output/12M_te.tab
```

```{bash}
head ../output/12M_te.tab
```


```{r}
testMte <- read.csv("../output/12M_te.tab", sep = "\t", header = FALSE)
```


```{r}
library(tidyverse)
```


```{r}

# create a vector of filenames with full path 
filenames <- list.files(path = "../../../../8TB_HDD_01/sr320/github/ceabigr/output", pattern = "mTE.out", full.names = TRUE)  

b <- data.frame()   # create empty dataframe to be populated with mean & median summary stats for each feature within a sample 

for (i in 1:length(filenames)) {  
  print(filenames[i])  # print out file location and name 
  testMte <- read.csv(file = filenames[i], sep = "\t", header = FALSE) # read in each sample data 
    # summarize methylation data per feature, mean & median 
    group12 <- testMte %>% group_by(V13,V5, V8, V9) %>%   
    summarize(avg = mean(V4, na.rm=TRUE), median=median(V4, na.rm=TRUE)) %>%

    # add new column with sample name 
    mutate(sample=gsub("../../../../8TB_HDD_01/sr320/github/ceabigr/output/", "", filenames[i]))
  b <- rbind(b, group12)
}
colnames(b) <- c("feature_name","mean_meth", "median_meth", "sample")

write.table(b, file = "../../../../8TB_HDD_01/sr320/github/ceabigr/output/transposon_summary_allsamples.txt",quote = F, row.names = F, sep = "\t")


```

```{r}
filenames <- list.files(path = "../output", pattern = "mTE.out", full.names = TRUE)
list.files(path = "../output", pattern = "mTE.out", full.names = TRUE)


for(i in 1:10) {                                           # Head of for-loop
 
  x1 <- i^2                                                # Code block
  print(x1)                                                # Print results
}
```






# CODE TO GET METHYLATION INTERSECT WITH TE

```{bash}
cd ../data
FILES=$(ls *bedgraph)

for file in ${FILES}
do
   NAME=$(echo ${file} | awk -F "_" '{print $1}')
   echo ${NAME}

  /home/shared/bedtools2/bin/intersectBed \
  -wb \
  -a ${NAME}_R1_val_1_10x.bedgraph \
  -b ../genome-features/C_virginica-3.0_TE-all.gff \
  > ../output/${NAME}mTE.out
done  



```
```{bash}
head ../output/9MmTE.out
```


# Taking TE methylation at each Loci and getting Mean methylation


```{r}
list.files(path = "../output", pattern = "mTE.out", full.names = TRUE)


```


```{r}
filenames <- list.files(path = "../output", pattern = "mTE.out", full.names = TRUE)

```


# put into a list
l <- list(df1, df2)

# iterate over the list with a for loop
for(i in 1:length(l)){
  l[[i]]$new_column <- l[[i]]$x + l[[i]]$y
}

list.files(path = "data", pattern = "csv", full.names = TRUE)

```{r}

for (f in filenames) {
  print(f)
}


```






