---
title: "33-isoform-count_meth"
output: html_document
---

```{r}
library(tidyverse)
```


Will try to take each individual samples isoform count and correlate with gene methylation.


```{r}
count_all <- read.csv("https://raw.githubusercontent.com/epigeneticstoocean/2018_L18-adult-methylation/main/analyses/transcript_counts_per_gene_per_sample.csv")
```

Need to do pivot longer
```{r}
count_all %>%
  pivot_longer(cols = c(2:27)) %>%
  mutate(geneID = paste("gene", gene_name, sep = '-'))
```
```{r}
count_all %>%
  pivot_longer(cols = c(2:27)) %>%
  mutate(geneID = paste("gene", gene_name, sep = '-')) %>%
  separate(name, into = c("tr", "sample"), sep = 19) %>%
  mutate(art = paste(geneID, sample, sep = '_' )) %>%
  rename(isoform_count = value) %>%
  select(art, isoform_count)
```


```{r}
ic <- count_all %>%
  pivot_longer(cols = c(2:27)) %>%
  mutate(geneID = paste("gene", gene_name, sep = '-')) %>%
  separate(name, into = c("tr", "sample"), sep = 19) %>%
  mutate(art = paste(geneID, sample, sep = '_' )) %>%
  rename(isoform_count = value) %>%
  select(art, isoform_count)
```

# Correlation of differences in max transcripts per gene with average gene methylation

NOTE: Incomplete as of 20221110 - SJW
```{r}
df <-diffs.max.tx.coordinates.list[["diffs.max.transcripts_per_gene.controls.vs.exposed"]] %>% select(name, difference) %>% arrange(name)


# Calculate mean methylation and sort by chromosome
meth.mean.gene <- methylation_coordinates %>%
  select(starts_with(c("chr", "name", "start", "end", all))) %>% 
  mutate(
    methylation_mean = 
      rowMeans(
        select(
          ., 
          contains(
            all, ignore.case = FALSE
            )
          ), na.rm = TRUE
        )
    ) %>%
  select(ends_with(c("name", "mean"))) %>% # Retain/reorder rows for joining
  arrange(name) # Sort by gene name

merged.meth.diffs.max.tx <- merge(df, meth.mean.gene, by = "name", all.x = TRUE)

plot(merged.meth.diffs.max.tx$methylation_mean, merged.meth.diffs.max.tx$difference, pch = 19, col = "lightblue");abline(lm(merged.meth.diffs.max.tx$difference ~ merged.meth.diffs.max.tx$methylation_mean), col = "red", lwd = 3)
```
