---
title: "33-isoform-count_meth"
output: html_document
---

```{r}
library(tidyverse)
```


Will try to take each individual samples isoform count and correlate with gene methylation.


```{r}
count_all <- read.csv("https://raw.githubusercontent.com/epigeneticstoocean/2018_L18-adult-methylation/main/analyses/transcript_counts_per_gene_per_sample.csv")
```

Need to do pivot longer
```{r}
count_all %>%
  pivot_longer(cols = c(2:27)) %>%
  mutate(geneID = paste("gene", gene_name, sep = '-'))
```
```{r}
count_all %>%
  pivot_longer(cols = c(2:27)) %>%
  mutate(geneID = paste("gene", gene_name, sep = '-')) %>%
  separate(name, into = c("tr", "sample"), sep = 19) %>%
  mutate(art = paste(geneID, sample, sep = '_' )) %>%
  rename(isoform_count = value) %>%
  select(art, isoform_count)
```


```{r}
ic <- count_all %>%
  pivot_longer(cols = c(2:27)) %>%
  mutate(geneID = paste("gene", gene_name, sep = '-')) %>%
  separate(name, into = c("tr", "sample"), sep = 19) %>%
  mutate(art = paste(geneID, sample, sep = '_' )) %>%
  rename(isoform_count = value) %>%
  select(art, isoform_count)
```
