---
title: "42-predominant-isoform"
output: html_document
---

# Set up R Markdown document

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = normalizePath("../output/42-predominant-isoform/")) #Set root directory
```

```{r}
getwd()
```

# Install packages

```{r}
# if ("tidyverse" %in% rownames(installed.packages()) == 'FALSE') install.packages('tidyverse')
# if ("ggpubr" %in% rownames(installed.packages()) == 'FALSE') install.packages('ggpubr')
# if ("lme4" %in% rownames(installed.packages()) == 'FALSE') install.packages('lme4')
# if ("data.table" %in% rownames(installed.packages()) == 'FALSE') install.packages('data.table')
# if ("RColorBrewer" %in% rownames(installed.packages()) == 'FALSE') install.packages('RColorBrewer')
# if ("patchwork" %in% rownames(installed.packages()) == 'FALSE') install.packages('patchwork')
```

```{r}
require(tidyverse)
require(ggpubr)
require(lme4)
require(data.table)
require(RColorBrewer)
require(patchwork)
```

```{r}
sessionInfo()
```

<https://raw.githubusercontent.com/epigeneticstoocean/2018_L18-adult-methylation/main/data/whole_tx_table.csv>

# Set variables

```{r set-variables}
# Vectors for subsetting samples by different groups
controls <- c("S13M", "S16F", "S19F", "S39F", "S44F", "S52F", "S53F", "S54F", "S64M", "S6M", "S76F", "S7M")
exposed <- c("S12M", "S22F", "S23M", "S29F", "S31M", "S35F", "S36F", "S3F", "S41F", "S48M", "S50F", "S59M", "S77F", "S9M")
controls_males <- c("S13M", "S64M", "S6M", "S7M")
exposed_males <- c("S12M", "S23M", "S31M", "S48M", "S59M", "S9M")
controls_females <- c("S16F", "S19F", "S39F", "S44F", "S52F", "S53F", "S54F", "S76F")
exposed_females <- c("S22F", "S29F", "S35F", "S36F", "S3F", "S41F", "S50F", "S77F")
```

```{r}
tx_exp <- read.csv("../../data/whole_tx_table.csv")
```

```{r}
tx_exp
```

```{r}
tx_exp %>%
  select(starts_with(c("gene_name", "t_name", "FPKM"))) %>%
  pivot_longer(cols = c(3:28)) %>%
  group_by(gene_name, t_name) %>%
  summarise(Predom_exp = mean(value))
```

```{r}
tx_exp %>%
  select(starts_with(c("gene_name", "t_name", "FPKM"))) %>%
  pivot_longer(cols = c(3:28)) %>%
  group_by(gene_name, t_name) %>%
  summarise(Predom_exp = mean(value)) %>%
  group_by(gene_name) %>%
  slice(which.max(Predom_exp))
```

```{r}
predom_all <- tx_exp %>%
  select(starts_with(c("gene_name", "t_name", "FPKM"))) %>%
  pivot_longer(cols = c(3:28)) %>%
  group_by(gene_name, t_name) %>%
  summarise(Predom_exp = mean(value)) %>%
  group_by(gene_name) %>%
  slice(which.max(Predom_exp))
```

```{r}
write.table(predom_all, file = "../output/42-predominant-isoform/predom_iso-all.txt",quote = F, row.names = F, sep = "\t")
```

Now can we do it for each "treatment" Let first just try to separate males and females

# Separate by sex

## Females

```{r}
fem_exp <- tx_exp %>%
  select(starts_with(c("gene_name", "t_name", "FPKM"))) %>%
  select(ends_with(c("name", "name", "F"))) %>%
  pivot_longer(cols = c(3:18)) %>%
  group_by(gene_name, t_name) %>%
  summarise(Predom_exp = mean(value)) %>%
  group_by(gene_name) %>%
  slice(which.max(Predom_exp))
```

```{r}
write.table(fem_exp, file = "../output/42-predominant-isoform/predom_iso-FEMALE.txt",quote = F, row.names = F, sep = "\t")
```

## Males

```{r}
male_exp <- tx_exp %>%
  select(starts_with(c("gene_name", "t_name", "FPKM"))) %>%
  select(ends_with(c("name", "name", "M"))) %>%
  pivot_longer(cols = c(3:12)) %>%
  group_by(gene_name, t_name) %>%
  summarise(Predom_exp = mean(value)) %>%
  group_by(gene_name) %>%
  slice(which.max(Predom_exp))
```

```{r}
tx_exp %>%
  select(starts_with(c("gene_name", "t_name", "FPKM"))) %>%
  select(ends_with(c("name", "name", "M"))) %>%
  pivot_longer(cols = c(3:12)) %>%
  group_by(gene_name, t_name) %>%
  summarise(Predom_exp = mean(value)) %>%
  group_by(gene_name) %>%
  slice(which.max(Predom_exp)) %>%
  write.table(file = "../output/42-predominant-isoform/predom_iso-MALE.txt",quote = F, row.names = F, sep = "\t")
```

Lets join male and female.

```{r}
mpre <- tx_exp %>%
  select(starts_with(c("gene_name", "t_name", "FPKM"))) %>%
  select(ends_with(c("name", "name", "M"))) %>%
  pivot_longer(cols = c(3:12)) %>%
  group_by(gene_name, t_name) %>%
  summarise(Predom_exp = mean(value)) %>%
  group_by(gene_name) %>%
  slice(which.max(Predom_exp))
```

```{r}
fpre <- tx_exp %>%
  select(starts_with(c("gene_name", "t_name", "FPKM"))) %>%
  select(ends_with(c("name", "name", "F"))) %>%
  pivot_longer(cols = c(3:18)) %>%
  group_by(gene_name, t_name) %>%
  summarise(Predom_exp = mean(value)) %>%
  group_by(gene_name) %>%
  slice(which.max(Predom_exp))
```

```{r}
full_join(mpre, fpre, by = "gene_name") %>%
  filter(t_name.x!=t_name.y) %>%
  filter(abs(Predom_exp.x - Predom_exp.y) > 100)
```

```{r}
anti_join(mpre, fpre, by = "t_name")
```

```{r}
anti_join(fpre, mpre, by = "t_name")
```

# Separate by sex and treatment

**NOTE: SAM DID THIS IN A JUPYTER NOTEBOOK: <https://nbviewer.org/github/sr320/ceabigr/blob/main/code/42-predominant_isoform-female_male.ipynb>**

## Males

### Exposed

```{r exposed-males-predominant-isoforms}
males_exposed <- tx_exp %>%
  select(starts_with(c("gene_name", "t_name", "FPKM"))) %>%
  select(ends_with(c("name", "name", exposed_males))) %>%
  pivot_longer(cols = c(3:length(exposed_males) + 2)) %>% # Adds length of vector "exposed males" + "name" + "name" from previous line
  group_by(gene_name, t_name) %>%
  summarise(Predom_exp = mean(value)) %>%
  group_by(gene_name) %>%
  slice(which.max(Predom_exp))
```

#### Write to file
```{r} 
write.table(file = "../output/42-predominant-isoform/predom_iso-exposed-MALE.txt",quote = F, row.names = F, sep = "\t")
```

### Control

```{r controls-males-predominant-isoforms}
mco <- tx_exp %>%
  select(starts_with(c("gene_name", "t_name", "FPKM"))) %>%
  select(ends_with(c("name", "name", controls_males))) %>%
  pivot_longer(cols = c(3:length(controls_males) + 2)) %>%
  group_by(gene_name, t_name) %>%
  summarise(Predom_exp = mean(value)) %>%
  group_by(gene_name) %>%
  slice(which.max(Predom_exp))
head(mco)
```



```{r}
write.table(mco, file = "../output/42-predominant-isoform/predom_iso-controls-MALE.txt",quote = F, row.names = F, sep = "\t")
```

## Females

### Exposed

```{r exposed-females-predominant-isoforms}
females_exposed <- tx_exp %>%
  select(starts_with(c("gene_name", "t_name", "FPKM"))) %>%
  select(ends_with(c("name", "name", exposed_females))) %>%
  pivot_longer(cols = c(3:length(exposed_females) + 2)) %>% # Adds length of vector "exposed males" + "name" + "name" from previous line
  group_by(gene_name, t_name) %>%
  summarise(Predom_exp = mean(value)) %>%
  group_by(gene_name) %>%
  slice(which.max(Predom_exp))
```

#### Write to file
```{r} 
write.table(females_exposed, file = "../output/42-predominant-isoform/predom_iso-exposed-FEMALE.txt",quote = F, row.names = F, sep = "\t")
```

### Control

```{r controls-females-predominant-isoforms}
females_controls <- tx_exp %>%
  select(starts_with(c("gene_name", "t_name", "FPKM"))) %>%
  select(ends_with(c("name", "name", controls_females))) %>%
  pivot_longer(cols = c(3:length(controls_females) + 2)) %>%
  group_by(gene_name, t_name) %>%
  summarise(Predom_exp = mean(value)) %>%
  group_by(gene_name) %>%
  slice(which.max(Predom_exp))
```

#### Write to file
```{r}
write.table(file = "../output/42-predominant-isoform/predom_iso-controls-FEMALE.txt",quote = F, row.names = F, sep = "\t")
```

```{r}
full_join(mco, mex, by = "gene_name") %>%
  filter(t_name.x!=t_name.y) %>%
  filter(abs(Predom_exp.x - Predom_exp.y) > 10) %>%
  head()
```

# Join Female and Male Controls/Exposed

## Rename `t_name` columns

```{r rename-columns}
# Rename t_name columns in each data frame
names(females_controls)[names(females_controls) == "t_name"] <- "females_controls_predom_isoform"
names(females_exposed)[names(females_exposed) == "t_name"] <- "females_exposed_predom_isoform"
names(males_exposed)[names(males_exposed) == "t_name"] <- "males_exposed_predom_isoform"
names(mco)[names(mco) == "t_name"] <- "males_controls_predom_isoform"
```

## Rename `Predom_exp` columns

```{r rename-predom-exp-columns}
# Rename t_name columns in each data frame
names(females_controls)[names(females_controls) == "Predom_exp"] <- "females_controls_predom_expression"
names(females_exposed)[names(females_exposed) == "Predom_exp"] <- "females_exposed_predom_expression"
names(males_exposed)[names(males_exposed) == "Predom_exp"] <- "males_exposed_predom_expression"
names(mco)[names(mco) == "Predom_exp"] <- "males_controls_predom_expression"
```

## Create single table

```{r join-female-male-controls-exposed}
predom_fcoe_mcoe <- predom_all %>%
  left_join(females_controls, by = "gene_name") %>%
  left_join(females_exposed, by = "gene_name") %>%
  left_join(mco, by = "gene_name") %>% 
  left_join(males_exposed, by = "gene_name") %>%
  dplyr::select(gene_name, ends_with("predom_isoform") | ends_with("predom_expression"))

str(predom_fcoe_mcoe)
```

## Write to file
```{r}
write.csv(predom_fcoe_mcoe,
          file = "predom_iso-FEMALE-and-MALE-controls-exposed.csv",
          quote = FALSE,
          row.names = FALSE)
```

# Analyse data

Based on the files from [Sam's Jupyter notebook](https://nbviewer.org/github/sr320/ceabigr/blob/main/code/42-predominant_isoform-female_male.ipynb), I will address the following questions:

-   How often does the predominant isoform change under OA conditions? Is this more than expected by random chance?
-   Is there a connection between predominant isoform changes, gene expression, and gene methylation?

I will do this for each sex separately.

## Import data

```{r}
metadata <- read.csv("../../data/adult-meta.csv", header = TRUE) %>%
  dplyr::rename(., sample = Sample.ID) %>%
  dplyr::rename(., treatment = Treatment) %>%
  mutate(., treatment = gsub(treatment, pattern = "Control", replacement = "control")) %>%
  mutate(., treatment = gsub(treatment, pattern = "Exposed", replacement = "exposed")) #Import metadata. Change column name and contents to be consistent throughout script
head(metadata)
```

### Isoform data

```{r}
predIsoFem <- read.delim("predom_iso-diff-controls_vs_exposed-female.tab", sep = "\t", header = FALSE, col.names = c("gene_name", "change"))
predIsoFem$gene_name <- str_trim(predIsoFem$gene_name, "right") #Remove extra spaces at end of gene name 
head(predIsoFem) #Confirm import
```

```{r}
predIsoMale <- read.delim("predom_iso-diff-controls_vs_exposed-male.tab", sep = "\t", header = FALSE, col.names = c("gene_name", "change"))
predIsoMale$gene_name <- str_trim(predIsoMale$gene_name, "right") #Remove extra spaces at end of gene name 
head(predIsoMale) #Confirm import
```

```{r}
totTranscripts <- read.delim("../34-transcript-counts/transcripts-counts-max_per_gene.tab", sep = "\t", header = FALSE, col.names = c("gene_name", "max_transcript")) #Import max transcript expressed per gene
head(totTranscripts)
```

### Methylation data

```{r}
geneMethylation <- read.csv("../40-gene-methylaiton.csv", header = TRUE, row.names = 2)
geneMethylationMod <- geneMethylation %>%
  select(., -1) #Remove extra column
head(geneMethylationMod)
```

```{r}
geneMethylationFem <- geneMethylationMod %>%
  filter(., grepl("F", rownames(.), fixed = TRUE)) %>% 
  t(.) %>%
  as.data.frame(.) %>%
  drop_na(.) %>%
  as.data.frame(.) %>% 
  rownames_to_column(., var = "gene_name") %>%
  pivot_longer(., !gene_name, names_to = "sample", values_to = "meth") %>%
  mutate(., gene_name = gsub(pattern = "gene.", replacement = "", gene_name)) %>%
  left_join(., metadata, by = "sample") %>%
  select(., c(gene_name, meth, treatment)) %>%
  dplyr::group_by(gene_name, treatment) %>%
  summarise(avg_meth = mean(meth, na.rm = TRUE)) %>%
  ungroup(.) #Filter female samples and reformat data. Remove excess information from gene name column. Add treatment information and average methylation for each gene by treatment.
head(geneMethylationFem) #Confirm formatting. Resultant df has one line per gene/treatment.
```

```{r}
geneMethylationMale <- geneMethylationMod %>%
  filter(., grepl("M", rownames(.), fixed = TRUE)) %>% 
  t(.) %>%
  as.data.frame(.) %>%
  drop_na(.) %>%
  as.data.frame(.) %>% 
  rownames_to_column(., var = "gene_name")  %>%
  pivot_longer(., !gene_name, names_to = "sample", values_to = "meth") %>%
  mutate(., gene_name = gsub(pattern = "gene.", replacement = "", gene_name)) %>%
  left_join(., metadata, by = "sample") %>%
  select(., c(gene_name, meth, treatment)) %>%
  dplyr::group_by(gene_name, treatment) %>%
  summarise(avg_meth = mean(meth, na.rm = TRUE)) %>%
  ungroup(.) #Filter male samples and reformat data. Remove excess information from gene name column. Add treatment information and average methylation for each gene by treatment.
head(geneMethylationMale) #Confirm formatting
```

### Expression data

```{r}
geneExpression <- read.csv("../../data/gene_fpkm.csv", header = TRUE, row.names = metadata$Sample.ID)
geneExpressionMod <- geneExpression %>%
  dplyr::select(., -c(1:6)) #Remove extra information to have just FPKM for each gene
```

```{r}
geneExpressionFem <- geneExpressionMod %>%
  filter(., grepl("F", rownames(.), fixed = TRUE)) %>%
  t(.) %>%
  as.data.frame(.) %>%
  drop_na(.) %>%
  as.data.frame(.) %>% 
  rownames_to_column(., var = "gene_name")  %>%
  pivot_longer(., !gene_name, names_to = "sample", values_to = "exp") %>% 
  mutate(., gene_name = gsub("gene.", "", gene_name))  %>%
  left_join(., metadata, by = "sample") %>%
  select(., c(gene_name, exp, treatment)) %>%
  dplyr::group_by(gene_name, treatment) %>%
  summarise(avg_exp = mean(exp, na.rm = TRUE)) %>%
  ungroup(.) #Filter female samples and reformat data. Remove excess information from gene name column. Add treatment information and average expression for each gene by treatment.
head(geneExpressionFem) #Confirm formatting
```

```{r}
geneExpressionMale <- geneExpressionMod %>%
  filter(., grepl("M", rownames(.), fixed = TRUE)) %>%
  t(.) %>%
  as.data.frame(.) %>%
  drop_na(.) %>%
  as.data.frame(.) %>% 
  rownames_to_column(., var = "gene_name")  %>%
  pivot_longer(., !gene_name, names_to = "sample", values_to = "exp") %>% 
  mutate(., gene_name = gsub("gene.", "", gene_name)) %>%
  left_join(., metadata, by = "sample") %>%
  select(., c(gene_name, exp, treatment)) %>%
  dplyr::group_by(gene_name, treatment) %>%
  summarise(avg_exp = mean(exp, na.rm = TRUE)) %>%
  ungroup(.) #Filter male samples and reformat data. Remove excess information from gene name column. Add treatment information and average expression for each gene by treatment.
head(geneExpressionMale) #Confirm formatting
```

## Females

### Load and format data

The file has data for all genes, but I want to remove genes that did not have any expression data. I also want to include the number of transcripts per gene. If there is only one transcript for the gene, I don't want to include it in the analysis.

```{r}
#Join predominant isoform data with expression data
#Keep rows where exp > 0
#Join with number of transcript information
#Keep rows where # transcripts > 1

predIsoFemFilt <- predIsoFem %>%
  inner_join(., fem_exp, by = "gene_name") %>%
  filter(., Predom_exp > 0) %>%
  inner_join(., totTranscripts, by = "gene_name") %>%
  filter(., max_transcript > 1)
head(predIsoFemFilt) #Confirm formatting
```

```{r}
cov_data <- fread("https://raw.githubusercontent.com/epigeneticstoocean/2018_L18-adult-methylation/main/analyses/genes_treatment_fpkm_CoV_df.csv")
Cvir_genom_feats <- fread("https://gannet.fish.washington.edu/Atumefaciens/20211209_cvir_gff-to-bed/20211209_cvir_GCF_002022765.2_genes.bed")
```

```{r}
colnames(cov_data) <- c("gene_ids", "M_control","F_control", "M_treated", "F_treated")

cov_data_l <- cov_data %>% pivot_longer(names_to = "sample",values_to = "CoV_FPKM", cols = 2:5)

Cvir_genom_feats$length <- Cvir_genom_feats$V3 - Cvir_genom_feats$V2 + 1

colnames(Cvir_genom_feats)[4] <- "gene_ids"

cov_data_l <- merge(cov_data_l, Cvir_genom_feats[,c("gene_ids", "length")], by = "gene_ids")

head(cov_data_l) #Confirm formatting
```

```{r}
cov_data_fem <- cov_data_l %>%
  mutate(., sex = case_when(sample == "M_treated" ~ "M",
                            sample == "M_control" ~ "M",
                            sample == "F_treated" ~ "F",
                            sample == "F_control" ~ "F")) %>%
  mutate(., treatment = case_when(sample == "M_treated" ~ "exposed",
                                  sample == "M_control" ~ "control",
                                  sample == "F_treated" ~ "exposed",
                                  sample == "F_control" ~ "control")) %>%
  filter(., sex == "F") %>%
  filter(., is.na(CoV_FPKM) == FALSE) %>%
  mutate(., gene_name = gsub("gene-", "", gene_ids)) %>%
  select(., 3:7) #Create sex and treatment columns based on existing sample column. Filter for female data and remove entries where CoV_FPKM = NA. Remove prefix from gene names and excess columns. 
nrow(cov_data_fem)
head(cov_data_fem)
```

```{r}
methCovDataFem <- read.csv("../40-gene-methylation/controls.females_v_exposed.females.CoV-mean-methylation.csv", header = TRUE) %>%
  select(., -abs.delta.CoV) %>%
  dplyr::rename(., gene_name = gene.id) %>%
  mutate(., gene_name = gsub("gene-", "", gene_name)) %>%
  pivot_longer(., cols = 2:3, names_to = "treatment", values_to = "CoV_Meth") %>%
  mutate(., treatment = gsub(treatment, pattern = "s.females.mean.methylation.CoV", replacement = "")) %>%
  mutate(., treatment = gsub(treatment, pattern = ".females.mean.methylation.CoV", replacement = "")) %>%
  filter(., is.na(CoV_Meth) == FALSE) #Import CV data for female oysters. Remove deltaCV column and rename gene.id column, Remove prefix from gene names. Pivot table longer such that each row has data for either exposed or control oysters. Remove superflous information in treatment identification. Remove rows where CoV_Meth â‰  NA
nrow(methCovDataFem)
head(methCovDataFem)
```

```{r}
femMethExpCV <- inner_join(femMethExp, cov_data_fem, by = c("gene_name", "treatment")) %>%
  inner_join(., methCovDataFem, by = c("gene_name", "treatment")) %>%
  select(., -c(t_name, sex)) %>%
  distinct(.) #Join CV information for expression and methylation with gene-level expression and methylation data for each sample.
nrow(femMethExpCV %>% select(gene_name) %>% distinct(.)) #9854 unique genes
head(femMethExpCV)
```

### Isoform change

```{r}
nrow(predIsoFemFilt) #38575 genes total
sum(predIsoFemFilt$change) #2252 with shift in predominant isoform
sum(predIsoFemFilt$change) / nrow(predIsoFemFilt) * 100 #5.837978% of genes have predominant isoforms that change under OA conditions
```

```{r}
predIsoFemTest <- prop.test(x = 2252, n = 38575, p = 0.50,
                            alternative = "two.sided",
                            conf.level = 0.95) #Conduct 1-sample z-test. 
predIsoFemTest
predIsoFemTest$statistic #X-squared = 30091.12
predIsoFemTest$p.value #p = 0. Isoform shifts happen significantly less than what would be expected by random chance
```

#### Annotate isoforms

```{r}
mRNATrack <- read.delim("../../genome-features/C_virginica-3.0-mRNA.gff", header = FALSE) %>%
  separate(., V9, into = c("V10", "V11"), sep = "ID=rna-") %>%
  separate(., V11, into = c("V12", "V13"), sep = ";Parent=") %>%
  separate(., V13, into = c("V14", "V15"), sep = ";Dbxref=") %>%
  separate(., V15, into = c("V16", "V17"), sep = ";product=") %>%
  separate(., V17, into = c("V18", "V19"), sep = ";transcript_id=") %>%
  select(., c("V12", "V14", "V18")) %>%
  rename(., transcript = V12, gene = V14, product = V18) #Winnow down mRNA track
mRNATrack$gene_name <- gsub("gene-", "", mRNATrack$gene) #Make gene naming consistent
head(mRNATrack) #Confirm changes
```

```{bash}
wget https://gannet.fish.washington.edu/spartina/paper-gonad-meth/analyses/2018-12-02-Gene-Enrichment-Analysis/_blast-GO-unfolded.sorted > ../genome-features/_blast-GO-unfolded.sorted
```

```{bash}
wget https://gannet.fish.washington.edu/spartina/project-gigas-oa-meth/output/GO-term-annotation/GO-GOslim.sorted > ../genome-features/GO-GOslim.sorted
```

```{r}
GOterms <- read.delim("../../genome-features/_blast-GO-unfolded.sorted", header = FALSE, col.names = c("GO", "transcript"))
head(GOterms)
```

```{r}
GOslim <- read.delim("../../genome-features/GO-GOslim.sorted", header = FALSE, col.names = c("GO", "GOterm", "GOslim", "process"))
head(GOslim)
```

```{r}
head(predIsoFemFilt)
```

```{r}
predIsoFemFiltAnnot <- inner_join(predIsoFemFilt, mRNATrack, by = "gene_name") %>%
  inner_join(., GOterms, by = "transcript") %>%
  inner_join(., GOslim, by = "GO")
head(predIsoFemFiltAnnot)
```

```{r}
predIsoFemFiltAnnotP <- predIsoFemFiltAnnot %>%
  filter(., process == "P") %>%
  filter(., change == 0)
#  filter(., change == 1)
as.data.frame(table(predIsoFemFiltAnnotP$GOslim))
```

```{r}
predIsoFemFiltAnnotF <- predIsoFemFiltAnnot %>%
  filter(., process == "F") %>%
  #    filter(., change == 0)
  filter(., change == 1)
as.data.frame(table(predIsoFemFiltAnnotF$GOslim))
```

### Isoforms and methylation

```{r}
head(geneMethylationFem)
head(geneExpressionFem)
```

```{r}
femMethExp <- inner_join(geneMethylationFem, geneExpressionFem, by = c("gene_name", "treatment")) %>%
  inner_join(., predIsoFemFilt, by = "gene_name") %>%
  select(., 1:6) %>%
  mutate(meth_bins = case_when(avg_meth < 10 ~ "0",
                               avg_meth >= 10 & avg_meth < 20 ~ "10",
                               avg_meth >= 20 & avg_meth < 30 ~ "20",
                               avg_meth >= 30 & avg_meth < 40 ~ "30",
                               avg_meth >= 40 & avg_meth < 50 ~ "40",
                               avg_meth >= 50 & avg_meth < 60 ~ "50",
                               avg_meth >= 60 & avg_meth < 70 ~ "60",
                               avg_meth >= 70 & avg_meth < 80 ~ "70",
                               avg_meth >= 80 & avg_meth < 90 ~ "80",
                               avg_meth >= 90 & avg_meth < 100 ~ "90",
                               avg_meth >= 100 ~ "100")) %>%
  mutate(exp_bins = case_when(avg_exp < 10 ~ "0",
                              avg_exp >= 10 & avg_exp < 20 ~ "10",
                              avg_exp >= 20 & avg_exp < 30 ~ "20",
                              avg_exp >= 30 & avg_exp < 40 ~ "30",
                              avg_exp >= 40 & avg_exp < 50 ~ "40",
                              avg_exp >= 50 & avg_exp < 60 ~ "50",
                              avg_exp >= 60 & avg_exp < 70 ~ "60",
                              avg_exp >= 70 & avg_exp < 80 ~ "70",
                              avg_exp >= 80 & avg_exp < 90 ~ "80",
                              avg_exp >= 90 & avg_exp < 100 ~ "90",
                              avg_exp >= 100 ~ "100"))
head(femMethExp)
```

#### XY Plots

```{r}
femMethExp %>%
  filter(., change == 1) %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = log(avg_exp)), colour = "#006D2C") +    
  labs(x = "Methylation (%)", y = "log FPKM") +
  geom_smooth(aes(x = avg_meth, y = log(avg_exp)), method = lm, colour = "grey20", linewidth = 2, na.rm = TRUE) +
  theme_minimal()
ggsave("fem-pred-isoform-meth-exp-xy.pdf", height = 8.5, width = 11)
```

```{r}
femMethExp %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = log(avg_exp)), colour = "#006D2C") +    
  labs(x = "Methylation (%)", y = "log FPKM") +
  geom_smooth(aes(x = avg_meth, y = log(avg_exp)), method = lm, colour = "grey20", linewidth = 2, na.rm = TRUE) +
  theme_minimal()
ggsave("fem-meth-exp-xy.pdf", height = 8.5, width = 11)
```

```{r}
femMethExp %>%
  filter(., avg_meth !=0 & avg_exp !=0) %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = log(avg_exp), colour = factor(change)), alpha = 0.3) +
  scale_color_manual(values = c("grey", "black")) +
  theme_minimal()
```

```{r}
femMethExp %>%
  filter(., avg_meth !=0 & avg_exp !=0) %>%
  filter(., change == 1) %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = log(avg_exp), colour = bins)) +
  stat_smooth(method = lm, aes(x = avg_meth, y = log(avg_exp))) +
  stat_regline_equation(aes(x = avg_meth, y = log(avg_exp), label = paste(..eq.label.., ..rr.label.., ..adj.rr.label.., sep = "~~~~")), formula = y ~ x, size = 5) +
  theme_minimal()

femMethExp %>%
  filter(., avg_meth !=0 & avg_exp !=0) %>%
  filter(., change == 0) %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = log(avg_exp), colour = bins)) +
  stat_smooth(method = lm, aes(x = avg_meth, y = log(avg_exp))) +
  stat_regline_equation(aes(x = avg_meth, y = log(avg_exp), label = paste(..eq.label.., ..rr.label.., ..adj.rr.label.., sep = "~~~~")), formula = y ~ x, size = 5) +
  theme_minimal()
```


```{r}
femMethExp %>%
  filter(., avg_meth !=0 & avg_exp !=0) %>%
  filter(., change == 1) %>%
  ggplot(.) + geom_density(aes(x = log(avg_exp), color = bins)) +
  theme_minimal()

femMethExp %>%
  filter(., avg_meth !=0 & avg_exp !=0) %>%
  filter(., change == 0) %>%
  ggplot(.) + geom_density(aes(x = log(avg_exp), color = bins)) +
  theme_minimal()
```

#### Model data

I want to understand if a predominant isoform shift due to pH treatment can be explained by a change in gene body methylation due to pH. I will also add gene length (the longer the gene, the more potential for different isoforms/alternative start or stop sites) and change in gene expression due to pH (higher expression = higher likelihood of picking up signal associated with isoform change).

if the predominant isoform shifted under low pH  ~ change gene body methylation (exposed - control) + change in gene expression (exposed - control) + log10(gene length).

```{r}
predIsoFemWide <- femMethExpCV %>%
  select(., gene_name, treatment, change, avg_meth, avg_exp, length) %>%
  pivot_wider(names_from = treatment, values_from = c(avg_meth, avg_exp)) %>%
  replace(is.na(.), 0) %>%
  mutate(., change_meth = avg_meth_exposed - avg_meth_control) %>%
  mutate(., change_exp = avg_exp_exposed - avg_exp_control) #Take data and select gene name, change, treatment, average methylation, average expression, and length columns. Pivot wider, appending treatment information to names of the columns. Replace all NA with 0 (no expression or methylation detected). Calculate the change in methylation and expression.
head(predIsoFemWide)
```

```{r}
predIsoFemModel <- glm(I(change == 1) ~ change_meth +
                         log10(length) + log2(change_exp + 1),
                       family = binomial(),
                       data = predIsoFemWide) #Create a binomial model, where 1 = the predominant isoform shifted when comparing low and control pH conditions. Use change in methylation, gene length, and change in gene expression as explanatory variables
summary(predIsoFemModel) #Only length and gene expression significant
```

```{r}
step(predIsoFemModel, test = "LRT")
```

```{r}
predIsoFemModel2 <- glm(I(change == 1) ~ log10(length) + log2(change_exp + 1),
                       family = binomial(),
                       data = predIsoFemWide) #Model identified by step
summary(predIsoFemModel2) #Only length and gene expression significant
```

```{r}
write.csv(broom::tidy(predIsoFemModel2), "fem-predo-isoform-model-output.csv", quote = FALSE, row.names = FALSE) #Save model output
```

### Transcriptional noise analysis

#### Plot data

```{r}
plotColors <- RColorBrewer::brewer.pal(9, "Greens")
```

```{r}
femMethExpCV %>%
  filter(., change == 1) %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = CoV_FPKM), colour = plotColors[8]) +    
  labs(x = "Methylation (%)", y = "CV FPKM") +
  scale_y_continuous(limits = c(0,3)) +
  geom_smooth(aes(x = avg_meth, y = CoV_FPKM), method = lm, colour = "grey20", linewidth = 2, na.rm = TRUE) +
  theme_classic()
ggsave("fem-pred-isoform-meth-CV-xy.pdf", height = 8.5, width = 11)
```

```{r}
femMethExpCV %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = CoV_FPKM), colour = plotColors[8]) +    
  labs(x = "Methylation (%)", y = "CV FPKM") +
  scale_y_continuous(limits = c(0,3)) +
  geom_smooth(aes(x = avg_meth, y = CoV_FPKM), method = lm, colour = "grey20", linewidth = 2, na.rm = TRUE) +
  theme_classic()
ggsave("fem-meth-CV-xy.pdf", height = 8.5, width = 11)
```

```{r}
femMethExpCV %>%
  filter(., change == 1) %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = CoV_Meth), colour = plotColors[8]) +    
  labs(x = "Methylation (%)", y = "CV Methylation") +
  scale_y_continuous(limits = c(0,3)) +
  geom_smooth(aes(x = avg_meth, y = CoV_Meth), method = lm, colour = "grey20", linewidth = 2, na.rm = TRUE) +
  theme_minimal()
ggsave("fem-pred-isoform-meth-CVMeth-xy.pdf", height = 8.5, width = 11)
```

```{r}
femMethExpCV %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = CoV_Meth), colour = plotColors[8]) +    
  labs(x = "Methylation (%)", y = "CV Methylation") +
  scale_y_continuous(limits = c(0,3)) +
  geom_smooth(aes(x = avg_meth, y = CoV_Meth), method = lm, colour = "grey20", linewidth = 2, na.rm = TRUE) +
  theme_minimal()
ggsave("fem-meth-CVMeth-xy.pdf", height = 8.5, width = 11)
```

#### Model data

Model needed based on Wu et al. 2020:

log10(transcriptional noise) ~ gene body methylation + log2(gene expression) + log10(gene length) + pH treatment

```{r}
transNoiseFem <- lm(data = femMethExpCV,
                    log10(CoV_FPKM) ~ avg_meth + log10(CoV_Meth) + I((log2(avg_exp + 1)^2)) + log2(avg_exp + 1) + log10(length) + as.factor(treatment) + as.factor(treatment)*avg_meth + as.factor(treatment)*I((log2(avg_exp + 1)^2)) + as.factor(treatment)* log2(avg_exp + 1)) #Model with quadratic relationship between expression and transcriptional noise
summary(transNoiseFem) #All significant
```

```{r}
transNoiseFem2 <- lm(data = femMethExpCV,
                     log10(CoV_FPKM) ~ avg_meth + log10(CoV_Meth) + log2(avg_exp + 1) + log10(length) + as.factor(treatment) + as.factor(treatment)*avg_meth + as.factor(treatment)*log2(avg_exp + 1)) #Try model without polynomial relationship between expression and transcriptional noise
summary(transNoiseFem2) #Length and meth*treatment not significant.
```

```{r}
transNoiseFem3 <- lm(data = femMethExpCV,
                     log10(CoV_FPKM) ~ avg_meth + log10(CoV_Meth) + log2(avg_exp + 1) + as.factor(treatment) + as.factor(treatment)*log2(avg_exp + 1)) #Remove insignificant terms from transNoiseFem2
summary(transNoiseFem3) #All terms significant
```

```{r}
AIC(transNoiseFem, transNoiseFem2, transNoiseFem3) #Compare AIC between models. The first model with the polynomial relationship between expression and transcriptional noise has the lowest AIC, so it will be used as the final model.
```

```{r}
write.csv(broom::tidy(transNoiseFem), "transcriptional-noise/fem-model-output.csv", quote = FALSE, row.names = FALSE) #Save model output
```

```{r}
hist(residuals(transNoiseFem)) #Residuals normally distributed

plot(fitted(transNoiseFem), residuals(transNoiseFem)) #Variance of residuals roughly consistent for all observations. Borderline heteroscedasticity but probably okay...?
abline(h = 0, lty = 2, col = "grey")
```

##### Calculate partial R<sup>2</sup>

Partial R<sup>2</sup>, or the coefficient of partial determination, can tell us how much of the variation in the response variable is explained by the explanatory variable. This can tell us how "important" one variable is in the model.

```{r}
anova(transNoiseFem)
```

```{r}
transNoiseFemVP <- data.frame("variable" = rownames(anova(transNoiseFem)),
                              "SumSq" = anova(transNoiseFem)[[2]],
                              "partialR2" = NA) #Create dataframe with variable names and sum of squares information from ANOVA output

for (i in 1:nrow(transNoiseFemVP)) {
  transNoiseFemVP$partialR2[i] <- transNoiseFemVP$SumSq[i] / sum(transNoiseFemVP$SumSq)
} #Calculate partial r2 by dividing sum of squares for each variable by total sum of squares (including residuals)
transNoiseFemVP #Confirm dataframe creation
```

```{r}
write.csv(transNoiseFemVP, "transcriptional-noise/fem-model-partialR.csv", quote = FALSE, row.names = FALSE)
```

#### Plot model findings

```{bash}
mkdir transcriptional-noise
mkdir transcriptional-noise/figures
```

```{r}
femMethExpCV %>%
  mutate(., exp_bins = factor(exp_bins, c("0", "10", "20", "30", "40", "50", "60", "70", "80", "90", "100"))) %>%
  ggplot(aes(x = exp_bins, y = log10(CoV_FPKM))) + geom_violin(fill = "grey80", width = 1) +  
  geom_boxplot(width = 0.1, colour = "grey20", alpha = 0.2) +
  labs(x = "FPKM", y = "Transcriptional Noise") +
  stat_summary(aes(group = 1), fun = median, geom = "line", col = "black", lty = 2) +
  facet_grid(. ~ treatment, labeller = labeller(treatment = c(control = "Control", exposed = "Exposed"))) +
  theme_classic() + theme(legend.position = "none",
                          strip.text.x = element_text(size = 10, face = "bold"),
                          strip.background = element_rect(linewidth = 0))
ggsave("transcriptional-noise/figures/fem-TN-geneExp.pdf", height = 8.5, width = 11)
```

```{r}
femMethExpCV %>%
  ggplot(aes(x = log2(avg_exp + 1), y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[5], plotColors[8]),
                      name = "Treatment",
                      labels = c("Control", "Exposed")) +
  scale_alpha_continuous(guide = "none") +
  labs(x = "FPKM", y = "Transcriptional Noise") +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  theme_classic()
ggsave("transcriptional-noise/figures/fem-TN-FPKM.pdf", height = 8.5, width = 11)
```

There is lower transcriptional noise noise with higher gene expression expression, with exposed female oysters showing lower transcriptional noise. Two potential explanations: 1) under low pH stress, there is a directed stress response such that multiple isoforms are not necessary or 2) the pattern is driven by gene function (housekeeping vs. environmental response gene).

I'll make the same plot, but facet by GO Slim term to see if functional differences are driving the relationship between gene expression and transcriptional noise.

```{r}
femMethExpCV %>%
  left_join(., mRNATrack, by = "gene_name") %>%
  left_join(., GOterms, by = "transcript") %>%
  left_join(., GOslim, by = "GO") %>%
  select(., -c(transcript, gene, product, GO, GOterm)) %>%
  filter(., process == "P") %>%
  select(., -process) %>%
  distinct(.) %>%
  ggplot(aes(x = log2(avg_exp + 1), y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[5], plotColors[8])) +
  labs(x = "FPKM", y = "Transcriptional Noise") +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 1.5, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 1.5, na.rm = TRUE) +
  facet_wrap(. ~ GOslim, ncol = 3) +
  theme_classic() + theme(legend.position = "none",
                          strip.background = element_rect(linewidth = 0))
ggsave("transcriptional-noise/figures/fem-TN-FPKM-GOslim.pdf", height = 8.5, width = 11)
```

```{r}
femMethExpCV %>%
  ggplot(aes(x = log2(avg_exp + 1), y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[5], plotColors[8]),
                      name = "Treatment",
                      labels = c("Control", "Exposed")) +
  scale_alpha_continuous(guide = "none") +
  labs(x = "FPKM", y = "Transcriptional Noise") +
  geom_smooth(data = (. %>% filter(treatment == "control")), formula = y ~ poly(x,2), colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), formula = y ~ poly(x,2), colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  theme_classic()
ggsave("transcriptional-noise/figures/fem-TN-FPKM-squared.pdf", height = 8.5, width = 11)
```

```{r}
femMethExpCV %>%
  mutate(., meth_bins = factor(meth_bins, c("0", "10", "20", "30", "40", "50", "60", "70", "80", "90", "100"))) %>%
  ggplot(aes(x = meth_bins, y = log10(CoV_FPKM))) + geom_violin(fill = "grey80", width = 1) +  
  geom_boxplot(width = 0.1, colour = "grey20", alpha = 0.2) +
  labs(x = "Methylation (%)", y = "Transcriptional Noise") +
  stat_summary(aes(group = 1), fun = median, geom = "line", col = "black", lty = 2) +
  facet_grid(. ~ treatment, labeller = labeller(treatment = c(control = "Control", exposed = "Exposed"))) +
  theme_classic() + theme(legend.position = "none",
                          strip.text.x = element_text(size = 10, face = "bold"),
                          strip.background = element_rect(linewidth = 0))
ggsave("transcriptional-noise/figures/fem-TN-geneMeth.pdf", height = 8.5, width = 11)
```

```{r}
femMethExpCV %>%
  ggplot(aes(x = avg_meth, y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[5], plotColors[8]),
                      name = "Treatment",
                      labels = c("Control", "Exposed")) +
  scale_alpha_continuous(guide = "none") +
  labs(x = "Methylation (%)", y = "Transcriptional Noise") +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  theme_classic()
ggsave("transcriptional-noise/figures/fem-TN-meth.pdf", height = 8.5, width = 11)
```

Increased methylation decreases transcriptional noise, with exposed female oysters still having lower overall transcriptional noise. This fits in with the idea that methylation reduces spurious transcription or alternative splicing.

I'll facet by GOSlim term for the rest of the plots too.

```{r}
femMethExpCV %>%
  left_join(., mRNATrack, by = "gene_name") %>%
  left_join(., GOterms, by = "transcript") %>%
  left_join(., GOslim, by = "GO") %>%
  select(., -c(transcript, gene, product, GO, GOterm)) %>%
  filter(., process == "P") %>%
  select(., -process) %>%
  distinct(.) %>%
  ggplot(aes(x = avg_meth, y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[5], plotColors[8]),
                      name = "Treatment",
                      labels = c("Control", "Exposed")) +
  scale_alpha_continuous(guide = "none") +
  labs(x = "Methylation (%)", y = "Transcriptional Noise") +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  facet_wrap(. ~ GOslim, ncol = 3) +
  theme_classic() + theme(legend.position = "none",
                          strip.background = element_rect(linewidth = 0))
ggsave("transcriptional-noise/figures/fem-TN-meth-GOslim.pdf", height = 8.5, width = 11)
```

```{r}
femMethExpCV %>%
  ggplot(aes(x = log10(CoV_Meth), y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[5], plotColors[8]),
                      name = "Treatment",
                      labels = c("Control", "Exposed")) +
  scale_alpha_continuous(guide = "none") +
  labs(x = "CV Methylation", y = "Transcriptional Noise") +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  theme_classic()
ggsave("transcriptional-noise/figures/fem-TN-CVmeth.pdf", height = 8.5, width = 11)
```

Variation in methylation increases transcriptional noise.

```{r}
femMethExpCV %>%
  left_join(., mRNATrack, by = "gene_name") %>%
  left_join(., GOterms, by = "transcript") %>%
  left_join(., GOslim, by = "GO") %>%
  select(., -c(transcript, gene, product, GO, GOterm)) %>%
  filter(., process == "P") %>%
  select(., -process) %>%
  distinct(.) %>%
  ggplot(aes(x = log10(CoV_Meth), y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[5], plotColors[8]),
                      name = "Treatment",
                      labels = c("Control", "Exposed")) +
  scale_alpha_continuous(guide = "none") +
  labs(x = "CV Methylation", y = "Transcriptional Noise") +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  facet_wrap(. ~ GOslim, ncol = 3) +
  theme_classic() + theme(legend.position = "none",
                          strip.background = element_rect(linewidth = 0))
ggsave("transcriptional-noise/figures/fem-TN-CVmeth-GOslim.pdf", height = 8.5, width = 11)
```

```{r}
femMethExpCV %>%
  mutate(., treatment = case_when(treatment == "control" ~ "Control",
                                  treatment == "exposed" ~ "Exposed")) %>%
  ggplot(aes(x = treatment, y = log10(CoV_FPKM))) + geom_violin(aes(fill = treatment)) +   
  scale_fill_manual(values = c(plotColors[5], plotColors[8]),
                    guide = "none") +
  scale_alpha_continuous(guide = "none") +
  geom_boxplot(width = 0.1, colour = "grey20", alpha = 0.2) +
  labs(x = "Treatment", y = "Transcriptional Noise") +
  stat_summary(aes(group = 1), fun = median, geom = "line", col = "black", lty = 2) +
  theme_classic()
ggsave("transcriptional-noise/figures/fem-TN-treatment.pdf", height = 8.5, width = 11)
```

```{r}
femMethExpCV %>%
  left_join(., mRNATrack, by = "gene_name") %>%
  left_join(., GOterms, by = "transcript") %>%
  left_join(., GOslim, by = "GO") %>%
  select(., -c(transcript, gene, product, GO, GOterm)) %>%
  filter(., process == "P") %>%
  select(., -process) %>%
  distinct(.) %>%
  mutate(., treatment = case_when(treatment == "control" ~ "Control",
                                  treatment == "exposed" ~ "Exposed")) %>%
  ggplot(aes(x = treatment, y = log10(CoV_FPKM))) + geom_violin(aes(fill = treatment)) +   
  scale_fill_manual(values = c(plotColors[5], plotColors[8]),
                    guide = "none") +
  scale_alpha_continuous(guide = "none") +
  geom_boxplot(width = 0.1, colour = "grey20", alpha = 0.2) +
  labs(x = "Treatment", y = "Transcriptional Noise") +
  stat_summary(aes(group = 1), fun = median, geom = "line", col = "black", lty = 2) +
  facet_wrap(. ~ GOslim, ncol = 3) +
  theme_classic() + theme(legend.position = "none",
                          strip.background = element_rect(linewidth = 0))
ggsave("transcriptional-noise/figures/fem-TN-treatment-GOslim.pdf", height = 8.5, width = 11)
```

```{r}
femMethExpCV %>%
  ggplot(aes(x = as.factor(change), y = log10(CoV_FPKM))) + geom_violin(aes(fill = as.factor(change))) +  
  scale_fill_manual(values = c(plotColors[5], plotColors[8]),
                    guide = "none") +
  scale_alpha_continuous(guide = "none") +
  geom_boxplot(width = 0.1, colour = "grey20", alpha = 0.2) +
  labs(x = "Predominant Isoform Shift", y = "Transcriptional Noise") +
  stat_summary(aes(group = 1), fun = median, geom = "line", col = "black", lty = 2) +
  theme_classic()
ggsave("transcriptional-noise/figures/fem-TN-predIso.pdf", height = 8.5, width = 11)
```

```{r}
femMethExpCV %>%
  left_join(., mRNATrack, by = "gene_name") %>%
  left_join(., GOterms, by = "transcript") %>%
  left_join(., GOslim, by = "GO") %>%
  select(., -c(transcript, gene, product, GO, GOterm)) %>%
  filter(., process == "P") %>%
  select(., -process) %>%
  distinct(.) %>%
  ggplot(aes(x = as.factor(change), y = log10(CoV_FPKM))) + geom_violin(aes(fill = as.factor(change))) +  
  scale_fill_manual(values = c(plotColors[5], plotColors[8]),
                    guide = "none") +
  scale_alpha_continuous(guide = "none") +
  geom_boxplot(width = 0.1, colour = "grey20", alpha = 0.2) +
  labs(x = "Predominant Isoform Shift", y = "Transcriptional Noise") +
  stat_summary(aes(group = 1), fun = median, geom = "line", col = "black", lty = 2) +
  facet_wrap(. ~ GOslim, ncol = 3) +
  theme_classic() + theme(legend.position = "none",
                          strip.background = element_rect(linewidth = 0))
ggsave("transcriptional-noise/figures/fem-TN-predIso-GOslim.pdf", height = 8.5, width = 11)
```

##### Create multipanel plot

```{r}
#Save individual plots

#A. Treatment

femTreatmentPlot <- femMethExpCV %>%
  mutate(., treatment = case_when(treatment == "control" ~ "Control",
                                  treatment == "exposed" ~ "Exposed")) %>%
  ggplot(aes(x = treatment, y = log10(CoV_FPKM))) + geom_violin(aes(fill = treatment)) +   
  scale_fill_manual(values = c(plotColors[4], plotColors[8]),
                    guide = "none") +
  scale_alpha_continuous(guide = "none") +
  geom_boxplot(width = 0.1, colour = "grey20", alpha = 0.2) +
  labs(x = "Treatment", y = "Transcriptional Noise", title = expression(paste("A.   ", beta, " = -8.62 x", 10^-3, ", t = -3.17, p = 0.002"))) +
  stat_summary(aes(group = 1), fun = median, geom = "line", col = "black", lty = 2) +
  theme_classic(base_size = 15)

#B. FPKM

femFPKMPlot <- femMethExpCV %>%
  ggplot(aes(x = log2(avg_exp + 1), y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[4], plotColors[8]),
                      guide = "none") +
  scale_alpha_continuous(guide = "none") +
  labs(x = "FPKM", y = "", title = expression(paste("B.   ", beta, " = -0.10, t = -51.19, p < ", 10^-15))) +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  theme_classic(base_size = 15)

#C. Methylation

femMethPlot <- femMethExpCV %>%
  ggplot(aes(x = avg_meth, y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[4], plotColors[8]),
                      guide = "none") +
  scale_alpha_continuous(guide = "none") +
  labs(x = "Methylation (%)", y = "Transcriptional Noise", title = expression(paste("C.   ", beta, " = -3.85 x ", 10^-3, ", t = -56.32, p < ", 10^-15))) +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  theme_classic(base_size = 15)

#D. CV Methylation

femCVMethPlot <- femMethExpCV %>%
  ggplot(aes(x = log10(CoV_Meth), y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[4], plotColors[8]),
                      guide = "none") +
  scale_alpha_continuous(guide = "none") +
  labs(x = "CV Methylation", y = "", title = expression(paste("D.   ", beta, " = 0.10, t = 35.44, p < ", 10^-15))) +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  theme_classic(base_size = 15)
```

```{r}
(femTreatmentPlot | femFPKMPlot) / (femMethPlot | femCVMethPlot) #Create patchwork plot
ggsave("transcriptional-noise/figures/fem-TN-multipanel.pdf", height = 8.5, width = 11)
```

## Males

### Format data

```{r}
#Join predominant isoform data with expression data
#Keep rows where exp > 0
#Join with number of transcript information
#Keep rows where # transcripts > 1

predIsoMaleFilt <- predIsoMale %>%
  inner_join(., male_exp, by = "gene_name") %>%
  filter(., Predom_exp > 0) %>%
  inner_join(., totTranscripts, by = "gene_name") %>%
  filter(., max_transcript > 1)
head(predIsoMaleFilt) #Confirm formatting
```

```{r}
cov_data_male <- cov_data_l %>%
  mutate(., sex = case_when(sample == "M_treated" ~ "M",
                            sample == "M_control" ~ "M",
                            sample == "F_treated" ~ "F",
                            sample == "F_control" ~ "F")) %>%
  mutate(., treatment = case_when(sample == "M_treated" ~ "exposed",
                                  sample == "M_control" ~ "control",
                                  sample == "F_treated" ~ "exposed",
                                  sample == "F_control" ~ "control")) %>%
  filter(., sex == "M") %>%
  filter(., is.na(CoV_FPKM) == FALSE) %>%
  mutate(., gene_name = gsub(pattern = "gene-", replacement = "", gene_ids)) %>%
  select(., 3:7) #Create sex and treatment columns based on existing sample column. Filter for female data and remove entries where CoV_FPKM = NA. Remove prefix from gene names and excess columns. 
nrow(cov_data_male)
head(cov_data_male)
```

```{r}
methCovDataMale <- read.csv("../40-gene-methylation/controls.males_v_exposed.males.CoV-mean-methylation.csv", header = TRUE) %>%
  select(., -abs.delta.CoV) %>%
  dplyr::rename(., gene_name = gene.id) %>%
  mutate(., gene_name = gsub("gene-", "", gene_name)) %>%
  pivot_longer(., cols = 2:3, names_to = "treatment", values_to = "CoV_Meth") %>%
  mutate(., treatment = gsub(treatment, pattern = "s.males.mean.methylation.CoV", replacement = "")) %>%
  mutate(., treatment = gsub(treatment, pattern = ".males.mean.methylation.CoV", replacement = "")) %>%
  filter(., is.na(CoV_Meth) == FALSE) #Import CV data for male oysters. Remove deltaCV column and rename gene.id column, Remove prefix from gene names. Pivot table longer such that each row has data for either exposed or control oysters. Remove superflous information in treatment identification. Remove rows where CoV_Meth â‰  NA
nrow(methCovDataMale)
head(methCovDataMale)
```

```{r}
maleMethExpCV <- inner_join(maleMethExp, cov_data_male, by = c("gene_name", "treatment")) %>%
  inner_join(., methCovDataMale, by = c("gene_name", "treatment")) %>%
  select(., -c(t_name, sex)) %>%
  distinct(.) #Join CV information for expression and methylation with gene-level expression and methylation data for each sample.
nrow(maleMethExpCV %>% select(gene_name) %>% distinct(.)) #9944 unique genes
head(maleMethExpCV)
```

### Isoform change

```{r}
nrow(predIsoMaleFilt) #10559 genes total
sum(predIsoMaleFilt$change) #1768 with shift in predominant isoform
sum(predIsoMaleFilt$change) / nrow(predIsoMaleFilt) * 100 #16.74401% of genes have predominant isoforms that change under OA conditions
```

```{r}
predIsoMaleTest <- prop.test(x = 1768, n = 10559, p = 0.50,
                             alternative = "two.sided",
                             conf.level = 0.95) #Conduct 1-sample z-test. 
predIsoMaleTest
predIsoMaleTest$statistic
predIsoMaleTest$p.value #Isoform shifts happen significantly less than what would be expected by random chance
```

#### Annotate isoforms

```{r}
head(predIsoMaleFilt)
nrow(predIsoMaleTest)
```

```{r}
predIsoMaleFiltAnnot <- inner_join(predIsoMaleFilt, mRNATrack, by = "gene_name") %>%
  inner_join(., GOterms, by = "transcript") %>%
  inner_join(., GOslim, by = "GO")
head(predIsoMaleFiltAnnot)
```

```{r}
predIsoMaleFiltAnnotP <- predIsoMaleFiltAnnot %>%
  filter(., process == "P") %>%
  filter(., change == 0)
#  filter(., change == 1)
as.data.frame(table(predIsoMaleFiltAnnotP$GOslim))
```

```{r}
predIsoMaleFiltAnnotF <- predIsoMaleFiltAnnot %>%
  filter(., process == "F") %>%
  #    filter(., change == 0)
  filter(., change == 1)
as.data.frame(table(predIsoMaleFiltAnnotF$GOslim))
```

### Isoforms and methylation

```{r}
head(geneMethylationMale)
head(geneExpressionMale)
```

```{r}
maleMethExp <- inner_join(geneMethylationMale, geneExpressionMale, by = c("gene_name", "treatment")) %>%
  inner_join(., predIsoMaleFilt, by = "gene_name") %>%
  select(., 1:6) %>%
  mutate(meth_bins = case_when(avg_meth < 10 ~ "0",
                               avg_meth >= 10 & avg_meth < 20 ~ "10",
                               avg_meth >= 20 & avg_meth < 30 ~ "20",
                               avg_meth >= 30 & avg_meth < 40 ~ "30",
                               avg_meth >= 40 & avg_meth < 50 ~ "40",
                               avg_meth >= 50 & avg_meth < 60 ~ "50",
                               avg_meth >= 60 & avg_meth < 70 ~ "60",
                               avg_meth >= 70 & avg_meth < 80 ~ "70",
                               avg_meth >= 80 & avg_meth < 90 ~ "80",
                               avg_meth >= 90 & avg_meth < 100 ~ "90",
                               avg_meth >= 100 ~ "100")) %>%
  mutate(exp_bins = case_when(avg_exp < 10 ~ "0",
                              avg_exp >= 10 & avg_exp < 20 ~ "10",
                              avg_exp >= 20 & avg_exp < 30 ~ "20",
                              avg_exp >= 30 & avg_exp < 40 ~ "30",
                              avg_exp >= 40 & avg_exp < 50 ~ "40",
                              avg_exp >= 50 & avg_exp < 60 ~ "50",
                              avg_exp >= 60 & avg_exp < 70 ~ "60",
                              avg_exp >= 70 & avg_exp < 80 ~ "70",
                              avg_exp >= 80 & avg_exp < 90 ~ "80",
                              avg_exp >= 90 & avg_exp < 100 ~ "90",
                              avg_exp >= 100 ~ "100"))
head(maleMethExp)
```

#### XY Plots

```{r}
maleMethExp %>%
  filter(., change == 1) %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = log(avg_exp)), colour = "#006D2C") +    
  labs(x = "Methylation (%)", y = "log FPKM") +
  geom_smooth(aes(x = avg_meth, y = log(avg_exp)), method = lm, colour = "grey20", linewidth = 2, na.rm = TRUE) +
  theme_minimal()
ggsave("male-pred-isoform-meth-exp-xy.pdf", height = 8.5, width = 11)
```

```{r}
maleMethExp %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = log(avg_exp)), colour = "#006D2C") +    
  labs(x = "Methylation (%)", y = "log FPKM") +
  geom_smooth(aes(x = avg_meth, y = log(avg_exp)), method = lm, colour = "grey20", linewidth = 2, na.rm = TRUE) +
  theme_minimal()
ggsave("male-meth-exp-xy.pdf", height = 8.5, width = 11)
```

```{r}
maleMethExp %>%
  filter(., avg_meth !=0 & avg_exp !=0) %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = log(avg_exp), colour = factor(change)), alpha = 0.3) +
  scale_color_manual(values = c("grey", "black")) +
  theme_minimal()
```

```{r}
maleMethExp %>%
  filter(., avg_meth !=0 & avg_exp !=0) %>%
  filter(., change == 1) %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = log(avg_exp), colour = bins)) +
  stat_smooth(method = lm, aes(x = avg_meth, y = log(avg_exp))) +
  stat_regline_equation(aes(x = avg_meth, y = log(avg_exp), label = paste(..eq.label.., ..rr.label.., ..adj.rr.label.., sep = "~~~~")), formula = y ~ x, size = 5) +
  theme_minimal()

maleMethExp %>%
  filter(., avg_meth !=0 & avg_exp !=0) %>%
  filter(., change == 0) %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = log(avg_exp), colour = bins)) +
  stat_smooth(method = lm, aes(x = avg_meth, y = log(avg_exp))) +
  stat_regline_equation(aes(x = avg_meth, y = log(avg_exp), label = paste(..eq.label.., ..rr.label.., ..adj.rr.label.., sep = "~~~~")), formula = y ~ x, size = 5) +
  theme_minimal()
```

```{r}
maleMethExp %>%
  filter(., avg_meth !=0 & avg_exp !=0) %>%
  filter(., change == 1) %>%
  ggplot(.) + geom_density(aes(x = log(avg_exp), color = bins)) +
  theme_minimal()

maleMethExp %>%
  filter(., avg_meth !=0 & avg_exp !=0) %>%
  filter(., change == 0) %>%
  ggplot(.) + geom_density(aes(x = log(avg_exp), color = bins)) +
  theme_minimal()
```

#### Model data

```{r}
predIsoMaleWide <- maleMethExpCV %>%
  select(., gene_name, treatment, change, avg_meth, avg_exp, length) %>%
  pivot_wider(names_from = treatment, values_from = c(avg_meth, avg_exp)) %>%
  replace(is.na(.), 0) %>%
  mutate(., change_meth = avg_meth_exposed - avg_meth_control) %>%
  mutate(., change_exp = avg_exp_exposed - avg_exp_control) #Take data and select gene name, change, treatment, average methylation, average expression, and length columns. Pivot wider, appending treatment information to names of the columns. Replace all NA with 0 (no expression or methylation detected). Calculate the change in methylation and expression.
head(predIsoMaleWide)
```

```{r}
predIsoMaleModel <- glm(I(change == 1) ~ change_meth +
                         log10(length) + log2(change_exp + 1),
                       family = binomial(),
                       data = predIsoMaleWide) #Create a binomial model, where 1 = the predominant isoform shifted when comparing low and control pH conditions. Use change in methylation, gene length, and change in gene expression as explanatory variables
summary(predIsoMaleModel) #Only length and gene expression significant
```

```{r}
step(predIsoMaleModel, test = "LRT")
```

The initial model is the most parsimonious based on the `step` results.

```{r}
write.csv(broom::tidy(predIsoMaleModel), "male-predo-isoform-model-output.csv", quote = FALSE, row.names = FALSE) #Save model output
```

### Transcriptional noise analysis

#### Plot data

```{r}
maleMethExpCV %>%
  filter(., change == 1) %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = CoV_FPKM), colour = "#006D2C") +    
  labs(x = "Methylation (%)", y = "CV FPKM") +
  scale_y_continuous(limits = c(0,3)) +
  geom_smooth(aes(x = avg_meth, y = CoV_FPKM), method = lm, colour = "grey20", linewidth = 2, na.rm = TRUE) +
  theme_classic()
ggsave("male-pred-isoform-meth-CV-xy.pdf", height = 8.5, width = 11)
```

```{r}
maleMethExpCV %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = CoV_FPKM), colour = "#006D2C") +    
  labs(x = "Methylation (%)", y = "CV FPKM") +
  scale_y_continuous(limits = c(0,3)) +
  geom_smooth(aes(x = avg_meth, y = CoV_FPKM), method = lm, colour = "grey20", linewidth = 2, na.rm = TRUE) +
  theme_classic()
ggsave("male-meth-CV-xy.pdf", height = 8.5, width = 11)
```

```{r}
maleMethExpCV %>%
  filter(., change == 1) %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = CoV_Meth), colour = "#006D2C") +    
  labs(x = "Methylation (%)", y = "CV Methylation") +
  scale_y_continuous(limits = c(0,3)) +
  geom_smooth(aes(x = avg_meth, y = CoV_Meth), method = lm, colour = "grey20", linewidth = 2, na.rm = TRUE) +
  theme_minimal()
ggsave("male-pred-isoform-meth-CVMeth-xy.pdf", height = 8.5, width = 11)
```

```{r}
maleMethExpCV %>%
  ggplot(.) + geom_point(aes(x = avg_meth, y = CoV_Meth), colour = "#006D2C") +    
  labs(x = "Methylation (%)", y = "CV Methylation") +
  scale_y_continuous(limits = c(0,3)) +
  geom_smooth(aes(x = avg_meth, y = CoV_Meth), method = lm, colour = "grey20", linewidth = 2, na.rm = TRUE) +
  theme_minimal()
ggsave("male-meth-CVMeth-xy.pdf", height = 8.5, width = 11)
```

#### Model data

```{r}
transNoiseMale <- lm(data = maleMethExpCV,
                     log10(CoV_FPKM) ~ avg_meth + log10(CoV_Meth) + I((log2(avg_exp + 1)^2)) + log2(avg_exp + 1) + log10(length) + as.factor(treatment) + as.factor(treatment)*avg_meth + as.factor(treatment)*I((log2(avg_exp + 1)^2)) + as.factor(treatment)* log2(avg_exp + 1)) #Model with polynomial relationship between expression and transcriptional noise
summary(transNoiseMale) #All terms significant
```

```{r}
transNoiseMale2 <- lm(data = maleMethExpCV,
                      log10(CoV_FPKM) ~ avg_meth + log10(CoV_Meth) + log2(avg_exp + 1) + log10(length) + as.factor(treatment) + as.factor(treatment)*avg_meth + as.factor(treatment)*log2(avg_exp + 1)) #Model without polynomial relationship between expression and transcriptional noise
summary(transNoiseMale2) #Length, treatment, and exp*treatment insignificant
```

```{r}
transNoiseMale3 <- lm(data = maleMethExpCV,
                      log10(CoV_FPKM) ~ avg_meth + log10(CoV_Meth) + log2(avg_exp + 1) + as.factor(treatment) + as.factor(treatment)*avg_meth) #Remove insignificant terms from transNoiseMale2 but keep treatment
summary(transNoiseMale3) #All terms significant except for treatment
```

```{r}
transNoiseMale4 <- lm(data = maleMethExpCV,
                      log10(CoV_FPKM) ~ avg_meth + log10(CoV_Meth) + log2(avg_exp + 1) + as.factor(treatment)*avg_meth) #Remove treatment from transNoiseMale3
summary(transNoiseMale4) #All terms significant
```

```{r}
AIC(transNoiseMale, transNoiseMale2, transNoiseMale3, transNoiseMale4) #Model with polynomial relationship has the lowest AIC, so that will be the final model
```

```{r}
write.csv(broom::tidy(transNoiseMale), "transcriptional-noise/male-model-output.csv", quote = FALSE, row.names = FALSE) #Save model output
```

```{r}
hist(residuals(transNoiseMale)) #Residuals normally distributed

plot(fitted(transNoiseMale2), residuals(transNoiseMale)) #Variance of residuals roughly consistent for all observations
abline(h = 0, lty = 2, col = "grey")
```

##### Calculate partial R2

Partial R<sup>2</sup>, or the coefficient of partial determination, can tell us how much of the variation in the response variable is explained by the explanatory variable. This can tell us how "important" one variable is in the model.

```{r}
anova(transNoiseMale)
```

```{r}
transNoiseMaleVP <- data.frame("variable" = rownames(anova(transNoiseMale)),
                               "SumSq" = anova(transNoiseMale)[[2]],
                               "partialR2" = NA) #Create dataframe with variable names and sum of squares information from ANOVA output

for (i in 1:nrow(transNoiseMaleVP)) {
  transNoiseMaleVP$partialR2[i] <- transNoiseMaleVP$SumSq[i] / sum(transNoiseMaleVP$SumSq)
} #Calculate partial r2 by dividing sum of squares for each variable by total sum of squares (including residuals)
transNoiseMaleVP #Confirm dataframe creation
```

```{r}
write.csv(transNoiseMaleVP, "transcriptional-noise/male-model-partialR.csv", quote = FALSE, row.names = FALSE)
```

#### Plot model findings

```{r}
maleMethExpCV %>%
  mutate(., exp_bins = factor(exp_bins, c("0", "10", "20", "30", "40", "50", "60", "70", "80", "90", "100"))) %>%
  ggplot(aes(x = exp_bins, y = log10(CoV_FPKM))) + geom_violin(fill = "grey80", width = 1) +  
  geom_boxplot(width = 0.1, colour = "grey20", alpha = 0.2) +
  labs(x = "FPKM", y = "Transcriptional Noise") +
  stat_summary(aes(group = 1), fun = median, geom = "line", col = "black", lty = 2) +
  facet_grid(. ~ treatment, labeller = labeller(treatment = c(control = "Control", exposed = "Exposed"))) +
  theme_classic() + theme(legend.position = "none",
                          strip.text.x = element_text(size = 10, face = "bold"),
                          strip.background = element_rect(linewidth = 0))
ggsave("transcriptional-noise/figures/male-TN-geneExp.pdf", height = 8.5, width = 11)
```

```{r}
maleMethExpCV %>%
  ggplot(aes(x = log2(avg_exp + 1), y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[5], plotColors[8]),
                      name = "Treatment",
                      labels = c("Control", "Exposed")) +
  scale_alpha_continuous(guide = "none") +
  labs(x = "FPKM", y = "Transcriptional Noise") +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  theme_classic()
ggsave("transcriptional-noise/figures/male-TN-FPKM.pdf", height = 8.5, width = 11)
```

Looking at expression in bins suggests that transcriptional noise increases with FPKM, but looking at the linear model shows the opposite. Treatment has no influence on the relationship between FPKM and transcriptional noise.

```{r}
maleMethExpCV %>%
  left_join(., mRNATrack, by = "gene_name") %>%
  left_join(., GOterms, by = "transcript") %>%
  left_join(., GOslim, by = "GO") %>%
  select(., -c(transcript, gene, product, GO, GOterm)) %>%
  filter(., process == "P") %>%
  select(., -process) %>%
  distinct(.) %>%
  ggplot(aes(x = log2(avg_exp + 1), y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[5], plotColors[8]),
                      name = "Treatment",
                      labels = c("Control", "Exposed")) +
  scale_alpha_continuous(guide = "none") +
  labs(x = "FPKM", y = "Transcriptional Noise") +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  facet_wrap(. ~ GOslim, ncol = 3) +
  theme_classic() + theme(legend.position = "none",
                          strip.background = element_rect(linewidth = 0))
ggsave("transcriptional-noise/figures/male-TN-FPKM-GOslim.pdf", height = 8.5, width = 11)
```

```{r}
maleMethExpCV %>%
  ggplot(aes(x = log2(avg_exp + 1), y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[5], plotColors[8]),
                      name = "Treatment",
                      labels = c("Control", "Exposed")) +
  scale_alpha_continuous(guide = "none") +
  labs(x = "FPKM", y = "Transcriptional Noise") +
  geom_smooth(data = (. %>% filter(treatment == "control")), formula = y ~ poly(x,2), colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), formula = y ~ poly(x,2), colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  theme_classic()
ggsave("transcriptional-noise/figures/male-TN-FPKM-squared.pdf", height = 8.5, width = 11)
```


```{r}
maleMethExpCV %>%
  mutate(., meth_bins = factor(meth_bins, c("0", "10", "20", "30", "40", "50", "60", "70", "80", "90", "100"))) %>%
  ggplot(aes(x = meth_bins, y = log10(CoV_FPKM))) + geom_violin(fill = "grey80", width = 1) +  
  geom_boxplot(width = 0.1, colour = "grey20", alpha = 0.2) +
  labs(x = "Methylation (%)", y = "Transcriptional Noise") +
  stat_summary(aes(group = 1), fun = median, geom = "line", col = "black", lty = 2) +
  facet_grid(. ~ treatment, labeller = labeller(treatment = c(control = "Control", exposed = "Exposed"))) +
  theme_classic() + theme(legend.position = "none",
                          strip.text.x = element_text(size = 10, face = "bold"),
                          strip.background = element_rect(linewidth = 0))
ggsave("transcriptional-noise/figures/male-TN-geneMeth.pdf", height = 8.5, width = 11)
```

```{r}
maleMethExpCV %>%
  ggplot(aes(x = avg_meth, y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[5], plotColors[8]),
                      name = "Treatment",
                      labels = c("Control", "Exposed")) +
  scale_alpha_continuous(guide = "none") +
  labs(x = "Methylation (%)", y = "Transcriptional Noise") +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  theme_classic()
ggsave("transcriptional-noise/figures/male-TN-meth.pdf", height = 8.5, width = 11)
```

Increased methylation decreases transcriptional noise.

```{r}
maleMethExpCV %>%
  left_join(., mRNATrack, by = "gene_name") %>%
  left_join(., GOterms, by = "transcript") %>%
  left_join(., GOslim, by = "GO") %>%
  select(., -c(transcript, gene, product, GO, GOterm)) %>%
  filter(., process == "P") %>%
  select(., -process) %>%
  distinct(.) %>%
  ggplot(aes(x = avg_meth, y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[5], plotColors[8]),
                      name = "Treatment",
                      labels = c("Control", "Exposed")) +
  scale_alpha_continuous(guide = "none") +
  labs(x = "Methylation (%)", y = "Transcriptional Noise") +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  facet_wrap(. ~ GOslim, ncol = 3) +
  theme_classic() + theme(legend.position = "none",
                          strip.background = element_rect(linewidth = 0))
ggsave("transcriptional-noise/figures/male-TN-meth-GOslim.pdf", height = 8.5, width = 11)
```

```{r}
maleMethExpCV %>%
  ggplot(aes(x = log10(CoV_Meth), y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[5], plotColors[8]),
                      name = "Treatment",
                      labels = c("Control", "Exposed")) +
  scale_alpha_continuous(guide = "none") +
  labs(x = "CV Methylation", y = "Transcriptional Noise") +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  theme_classic()
ggsave("transcriptional-noise/figures/male-TN-CVmeth.pdf", height = 8.5, width = 11)
```

Variation in methylation increases transcriptional noise.

```{r}
maleMethExpCV %>%
  left_join(., mRNATrack, by = "gene_name") %>%
  left_join(., GOterms, by = "transcript") %>%
  left_join(., GOslim, by = "GO") %>%
  select(., -c(transcript, gene, product, GO, GOterm)) %>%
  filter(., process == "P") %>%
  select(., -process) %>%
  distinct(.) %>%
  ggplot(aes(x = log10(CoV_Meth), y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[5], plotColors[8]),
                      name = "Treatment",
                      labels = c("Control", "Exposed")) +
  scale_alpha_continuous(guide = "none") +
  labs(x = "CV Methylation", y = "Transcriptional Noise") +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  facet_wrap(. ~ GOslim, ncol = 3) +
  theme_classic() + theme(legend.position = "none",
                          strip.background = element_rect(linewidth = 0))
ggsave("transcriptional-noise/figures/male-TN-CVmeth-GOslim.pdf", height = 8.5, width = 11)
```

```{r}
maleMethExpCV %>%
  mutate(., treatment = case_when(treatment == "control" ~ "Control",
                                  treatment == "exposed" ~ "Exposed")) %>%
  ggplot(aes(x = treatment, y = log10(CoV_FPKM))) + geom_violin(aes(fill = treatment)) +   
  scale_fill_manual(values = c(plotColors[5], plotColors[8]),
                    guide = "none") +
  scale_alpha_continuous(guide = "none") +
  geom_boxplot(width = 0.1, colour = "grey20", alpha = 0.2) +
  labs(x = "Treatment", y = "Transcriptional Noise") +
  stat_summary(aes(group = 1), fun = median, geom = "line", col = "black", lty = 2) +
  theme_classic()
ggsave("transcriptional-noise/figures/male-TN-treatment.pdf", height = 8.5, width = 11)
```

```{r}
maleMethExpCV %>%
  left_join(., mRNATrack, by = "gene_name") %>%
  left_join(., GOterms, by = "transcript") %>%
  left_join(., GOslim, by = "GO") %>%
  select(., -c(transcript, gene, product, GO, GOterm)) %>%
  filter(., process == "P") %>%
  select(., -process) %>%
  distinct(.) %>%
  mutate(., treatment = case_when(treatment == "control" ~ "Control",
                                  treatment == "exposed" ~ "Exposed")) %>%
  ggplot(aes(x = treatment, y = log10(CoV_FPKM))) + geom_violin(aes(fill = treatment)) +   
  scale_fill_manual(values = c(plotColors[5], plotColors[8]),
                    guide = "none") +
  scale_alpha_continuous(guide = "none") +
  geom_boxplot(width = 0.1, colour = "grey20", alpha = 0.2) +
  labs(x = "Treatment", y = "Transcriptional Noise") +
  stat_summary(aes(group = 1), fun = median, geom = "line", col = "black", lty = 2) +
  facet_wrap(. ~ GOslim, ncol = 3) +
  theme_classic() + theme(legend.position = "none",
                          strip.background = element_rect(linewidth = 0))
ggsave("transcriptional-noise/figures/male-TN-treatment-GOslim.pdf", height = 8.5, width = 11)
```

```{r}
maleMethExpCV %>%
  ggplot(aes(x = as.factor(change), y = log10(CoV_FPKM))) + geom_violin(aes(fill = as.factor(change))) +  
  scale_fill_manual(values = c(plotColors[5], plotColors[8]),
                    guide = "none") +
  scale_alpha_continuous(guide = "none") +
  geom_boxplot(width = 0.1, colour = "grey20", alpha = 0.2) +
  labs(x = "Predominant Isoform Shift", y = "Transcriptional Noise") +
  stat_summary(aes(group = 1), fun = median, geom = "line", col = "black", lty = 2) +
  theme_classic()
ggsave("transcriptional-noise/figures/male-TN-predIso.pdf", height = 8.5, width = 11)
```

```{r}
maleMethExpCV %>%
  left_join(., mRNATrack, by = "gene_name") %>%
  left_join(., GOterms, by = "transcript") %>%
  left_join(., GOslim, by = "GO") %>%
  select(., -c(transcript, gene, product, GO, GOterm)) %>%
  filter(., process == "P") %>%
  select(., -process) %>%
  distinct(.) %>%
  ggplot(aes(x = as.factor(change), y = log10(CoV_FPKM))) + geom_violin(aes(fill = as.factor(change))) +  
  scale_fill_manual(values = c(plotColors[5], plotColors[8]),
                    guide = "none") +
  scale_alpha_continuous(guide = "none") +
  geom_boxplot(width = 0.1, colour = "grey20", alpha = 0.2) +
  labs(x = "Predominant Isoform Shift", y = "Transcriptional Noise") +
  stat_summary(aes(group = 1), fun = median, geom = "line", col = "black", lty = 2) +
  facet_wrap(. ~ GOslim, ncol = 3) +
  theme_classic() + theme(legend.position = "none",
                          strip.background = element_rect(linewidth = 0))
ggsave("transcriptional-noise/figures/male-TN-predIso-GOslim.pdf", height = 8.5, width = 11)
```

##### Create multipanel plot

```{r}
#Save individual plots

#A. Treatment

maleTreatmentPlot <- maleMethExpCV %>%
  mutate(., treatment = case_when(treatment == "control" ~ "Control",
                                  treatment == "exposed" ~ "Exposed")) %>%
  ggplot(aes(x = treatment, y = log10(CoV_FPKM))) + geom_violin(aes(fill = treatment)) +   
  scale_fill_manual(values = c(plotColors[4], plotColors[8]),
                    guide = "none") +
  scale_alpha_continuous(guide = "none") +
  geom_boxplot(width = 0.1, colour = "grey20", alpha = 0.2) +
  labs(x = "Treatment", y = "Transcriptional Noise", title = expression(paste("A.   ", beta, " = 0.15, t = 7.81, p = 6.19 x ", 10^-15))) +
  stat_summary(aes(group = 1), fun = median, geom = "line", col = "black", lty = 2) +
  theme_classic(base_size = 15)

#B. FPKM

maleFPKMPlot <- maleMethExpCV %>%
  ggplot(aes(x = log2(avg_exp + 1), y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[4], plotColors[8]),
                      guide = "none") +
  scale_alpha_continuous(guide = "none") +
  labs(x = "FPKM", y = "", title = expression(paste("B.   ", beta, " = -0.41, t = -47.87, p < ", 10^-15))) +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  theme_classic(base_size = 15)

#C. Methylation

maleMethPlot <- maleMethExpCV %>%
  ggplot(aes(x = avg_meth, y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[4], plotColors[8]),
                      guide = "none") +
  scale_alpha_continuous(guide = "none") +
  labs(x = "Methylation (%)", y = "Transcriptional Noise", title = expression(paste("C.   ", beta, " = -1.40 x ", 10^-3, ", t = -14.88, p < ", 10^-15))) +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  theme_classic(base_size = 15)

#D. CV Methylation

maleCVMethPlot <- maleMethExpCV %>%
  ggplot(aes(x = log10(CoV_Meth), y = log10(CoV_FPKM))) + geom_point(aes(colour = treatment, alpha = 0.5)) +    
  scale_colour_manual(values = c(plotColors[4], plotColors[8]),
                      guide = "none") +
  scale_alpha_continuous(guide = "none") +
  labs(x = "CV Methylation", y = "", title = expression(paste("D.   ", beta, " = 0.14, t = 27.46, p = < ", 10^-15))) +
  geom_smooth(data = (. %>% filter(treatment == "control")), method = lm, colour = plotColors[7], linewidth = 2, na.rm = TRUE) +
  geom_smooth(data = (. %>% filter(treatment == "exposed")), method = lm, colour = plotColors[9], linewidth = 2, na.rm = TRUE) +
  theme_classic(base_size = 15)
```

```{r}
(maleTreatmentPlot | maleFPKMPlot) / (maleMethPlot | maleCVMethPlot) #Create patchwork plot
ggsave("transcriptional-noise/figures/male-TN-multipanel.pdf", height = 8.5, width = 11)
```



