---
title: "78 asca meth-diff"
author: Steven Roberts
date: "`r format(Sys.time(), '%d %B, %Y')`" 
output: 
  html_document:
    theme: readable
    highlight: zenburn
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
    code_download: true
  # github_document:
  #   toc: true
  #   toc_depth: 3
  #   number_sections: true
  #   html_preview: true
---

```{r setup, include=FALSE}
#library(kableExtra)
# library(DESeq2)
# library(pheatmap)
# library(RColorBrewer)
# library(data.table)
#library(DT)
# library(Biostrings)
#library(methylKit)
library(tidyverse)

knitr::opts_chunk$set(
  echo = TRUE,         # Display code chunks
  eval = TRUE,         # Evaluate code chunks
  warning = FALSE,     # Hide warnings
  message = FALSE,     # Hide messages
  fig.width = 6,       # Set plot width in inches
  fig.height = 4,      # Set plot height in inches
  fig.align = "center" # Align plots to the center
)
```

Ariana created lists of genes were exon patterning was similar and DIFFERENT under OA conditions for both sexes.

The question becomes is there a clear difference in methylation patterns in these genes. 

This includes determining the preponderance in DMLs and

if gene body methylation differed.

## female where pattern differed.


```{r}
female_splice <- read.table("../output/77-asca-exon/females_splicing_treatments_gene_list.csv", col.names = c("name")) %>%
  mutate(desc = "f_diff")
```


female DMLs



```{r}
female_dml <- read.table("../output/16-DML-annotation/female_dml_geneID.txt") %>%
  mutate(desc = "fdml")
```



join

```{r}
fsp_dml <- female_splice %>%
  left_join(female_dml, by = c("name" = "V1"))
```

nothing


# GBM

```{r}
female_40_meth <- read_csv("../output/73-gene-methylation/female_40_meth.csv") %>%
  mutate(name = str_replace(name, "gene-", ""))
```


```{r}
f_splice_meth <- female_40_meth %>%
  left_join(female_splice, by = c("name" = "name")) %>%
  filter(desc == "f_diff")
```


# plot 


```{r}
ggplot(f_splice_meth, aes(x = average_control, y = average_exposed)) +
  geom_point() +  # This adds the points to the plot
  theme_minimal() 
```

## female no splice


```{r}
female_NOsplice <- read.table("../output/77-asca-exon/females_splicing_nodifference_gene_list.csv", col.names = c("name")) %>%
  mutate(desc = "f_same")
```


female DMLs



```{r}
female_dml <- read.table("../output/16-DML-annotation/female_dml_geneID.txt") %>%
  mutate(desc = "fdml")
```



join

```{r}
fNOsp_dml <- female_NOsplice %>%
  left_join(female_dml, by = c("name" = "V1"))
```

1 dml overlap


# GBM

```{r}
female_40_meth <- read_csv("../output/73-gene-methylation/female_40_meth.csv") %>%
  mutate(name = str_replace(name, "gene-", ""))
```


```{r}
f_NOsplice_meth <- female_40_meth %>%
  left_join(female_NOsplice, by = c("name" = "name")) %>%
  filter(desc == "f_same")
```


# plot 


```{r}
ggplot(f_NOsplice_meth, aes(x = average_control, y = average_exposed)) +
  geom_point() +  # This adds the points to the plot
  theme_minimal() 
```


# lets look at together on one plot

```{r}
ggplot() +
  geom_point(data = f_NOsplice_meth, aes(x = average_control, y = average_exposed), color = "blue") +
  geom_point(data = f_splice_meth, aes(x = average_control, y = average_exposed), color = "red") +
  theme_minimal() +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
  geom_abline(intercept = -5, slope = 1, linetype = "solid", color = "grey") + # Line below
  geom_abline(intercept = 5, slope = 1, linetype = "solid", color = "grey") + # Line above
  labs(x = "control", y = "exposed", title = "female methylation difference")

```

# Males

## males where pattern differed.


```{r}
male_splice <- read.table("../output/77-asca-exon/males_splicing_treatments_gene_list.csv", col.names = c("name")) %>%
  mutate(desc = "m_diff")
```


male DMLs



```{r}
male_dml <- read.table("../output/16-DML-annotation/male_dml_geneID.txt") %>%
  mutate(desc = "mdml")
```



join

```{r}
msp_dml <- male_splice %>%
  left_join(male_dml, by = c("name" = "V1"))

msp_dml
```




# GBM

```{r}
male_40_meth <- read_csv("../output/73-gene-methylation/male_40_meth.csv") %>%
  mutate(name = str_replace(name, "gene-", ""))
```


```{r}
m_splice_meth <- male_40_meth %>%
  left_join(male_splice, by = c("name" = "name")) %>%
  filter(desc == "m_diff")
```


# plot 


```{r}
ggplot(m_splice_meth, aes(x = average_control, y = average_exposed)) +
  geom_point() +  # This adds the points to the plot
  theme_minimal() 
```

## male no splice


```{r}
male_NOsplice <- read.table("../output/77-asca-exon/males_splicing_nodifference_gene_list.csv", col.names = c("name")) %>%
  mutate(desc = "m_same")
```


male DMLs





join

```{r}
mNOsp_dml <- male_NOsplice %>%
  left_join(male_dml, by = c("name" = "V1"))
```




# GBM

```{r}
male_40_meth <- read_csv("../output/73-gene-methylation/male_40_meth.csv") %>%
  mutate(name = str_replace(name, "gene-", ""))
```


```{r}
m_NOsplice_meth <- male_40_meth %>%
  left_join(male_NOsplice, by = c("name" = "name")) %>%
  filter(desc == "m_same")
```


# plot 


```{r}
ggplot(m_NOsplice_meth, aes(x = average_control, y = average_exposed)) +
  geom_point() +  # This adds the points to the plot
  theme_minimal() 
```


# lets look at together on one plot

```{r}
ggplot() +
  geom_point(data = m_NOsplice_meth, aes(x = average_control, y = average_exposed), color = "blue") +
  geom_point(data = m_splice_meth, aes(x = average_control, y = average_exposed), color = "red") +
  theme_minimal() +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
  geom_abline(intercept = -5, slope = 1, linetype = "solid", color = "grey") + # Line below
  geom_abline(intercept = 5, slope = 1, linetype = "solid", color = "grey") + # Line above
  labs(x = "control", y = "exposed", title = "male methylation difference")

```


