---
title: "Creation of comprehensive CSV"
author: "Sam White"
date: "3/1/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load libraries
```{r load-libraries}
library("tidyverse")
```

# Load transcript FPKMs
```{r load-transcript-FPKMs}
transcripts_fpkm <- read.csv("../data/fpkm-all_transcripts.csv")

head(transcripts_fpkm)
```
# Load gene FPKMS
```{r load-gene-FPKMs}
genes_fpkm <- read.csv("../data/whole_gx_table.csv")

head(genes_fpkm)
```
# Load gene methylation
```{r load-gene-methylation}
genes_methylation <- read.csv("../output/40-gene-methylation/40-gene-methylation.csv")

head(genes_methylation)
```

# Rename transcript columns/rows
```{r rename-transcript-columns-and-rows}
# Remove "FPKM." from all column names
colnames(transcripts_fpkm) <- gsub("FPKM.", "", colnames(transcripts_fpkm))

head(transcripts_fpkm)

# Add "t_id." to beginning ("^") of first column data
transcripts_fpkm$t_id <- sub("^", "t_id-", transcripts_fpkm$t_id)

head(transcripts_fpkm)

# Add ".FPKM" to end ("$") of "t_id"
transcripts_fpkm$t_id <- sub("$", ".FPKM", transcripts_fpkm$t_id)

head(transcripts_fpkm)
```

# Transpose transcript FPKM
```{r transpose-transcripts-fpkm}
# Transpose to have sample names as rows and transcript FPKM in columns
transposed.transcripts_fpkm <- as.data.frame(t(transcripts_fpkm))

head(transposed.transcripts_fpkm)

# Convert first row to column names
colnames(transposed.transcripts_fpkm) <- transposed.transcripts_fpkm[1,]
transposed.transcripts_fpkm <- transposed.transcripts_fpkm[-1,]

head(transposed.transcripts_fpkm)
```


# Rename gene columns/rows
```{r rename-transcript-columns-and-rows}
# Remove summary first row
genes_fpkm <- genes_fpkm[-1,]

head(genes_fpkm)

# Remove "FPKM." from all column names
colnames(genes_fpkm) <- gsub("FPKM.", "", colnames(genes_fpkm))

head(genes_fpkm)

# Add ".FPKM" to end ("$") of "name"
genes_fpkm$name <- sub("$", ".FPKM", genes_fpkm$name)

head(genes_fpkm)
```

# Transpose gene FPKM
```{r transpose-genes-fpkm}
# Transpose to have sample names as rows and transcript FPKM in columns
transposed.genes_fpkm <- as.data.frame(t(genes_fpkm))

head(transposed.genes_fpkm)

# Convert first row to column names
colnames(transposed.genes_fpkm) <- transposed.genes_fpkm[1,]
transposed.genes_fpkm <- transposed.genes_fpkm[-1,]

head(transposed.genes_fpkm)
```

# Rename methylation columns/rows
```{r rename-transcript-columns-and-rows}
# Remove extra first column
genes_methylation <- genes_methylation[,-1]

head(genes_methylation)

# Replace "gene." with "gene-" from all column names for consistency
colnames(genes_methylation) <- gsub("gene.", "gene-", colnames(genes_methylation))

head(genes_methylation)

# Add ".meth" to end ("$") of all columns
colnames(genes_methylation) <- sub("$", ".meth", colnames(genes_methylation))

head(genes_methylation)

# Convert first column to rownames
genes_methylation <- genes_methylation %>%
  remove_rownames%>%
  column_to_rownames(var="sample_id.meth")
```

# Transpose methylation FPKM
```{r transpose-genes-fpkm}
# Transpose to have sample names as rows and transcript FPKM in columns
transposed.genes_methylation <- as.data.frame(t(genes_methylation))

head(transposed.genes_methylation)

# Convert first row to column names
colnames(transposed.genes_methylation) <- transposed.genes_methylation[1,]
transposed.genes_methylation <- transposed.genes_methylation[-1,]

head(transposed.genes_methylation)
```
