---
title: "exon-methylation-expression-mixomics"
output: html_document
---

# Set up R Markdown document

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "../output/exon-methylation-expression-mixomics/") #Set root directory
```

#Install packages

```{r}
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
# BiocManager::install("mixOmics") #Install BiocManager and mixOmics
require(mixOmics)
require(tidyverse)
```

```{r}
sessionInfo()
```

# THINGS TO DO BEFORE YOU DO ANYTHING YOU FOOL:

PLSDA
tune parameters
run spls
extract VIPs from spls: value > 1

# Import and reformat data

```{r}
metadata <- read.csv("../RAnalysis/data/adult-meta.csv", header = TRUE) #Import metadata
metadata <- metadata %>%
  select(., Sample.ID, Treatment, Sex, TreatmentN) %>%
  rename(., sample_id = Sample.ID) #Retain necessary columns, rename sample_id
head(metadata) #Confirm formatting
```

## Exon expression data

```{r}
exonExpression <- read.csv("/home/shared/8TB_HDD_02/ashuff/Projects/ceabigr/output/AH_gene_coverage_exon.csv", header = TRUE, row.names = 1)
exonExpressionMod <- exonExpression %>%
  unite(., exonID, 1:3) %>%
  select(., exonID, genes, sample_id) %>%
  distinct(.) %>%
  slice(., -212243) %>%
  spread(., key = exonID, value = genes, fill = NA) #Create exonID column with chr-start-end, select only exonID, gene_cov, and sample_id information, filter unique rows, spread data for mixOmics formatting requirements
head(exonExpressionMod) #Confirm formatting
ncol(exonExpressionMod) #Count number of exons with data
```

### Sex-specific data

```{r}
exonExpressionFem <- exonExpressionMod %>%
  filter(., grepl("F", sample_id, fixed = TRUE)) %>%
  select(where(~sum(!is.na(.)) > 0)) %>%
  column_to_rownames(., var = "sample_id") #Filter female samples only, remove exons with missing methylation data (NA) for any sample, and make sample IDs rownames
head(exonExpressionFem) #Confirm formatting
sum(is.na(exonExpressionFem)) #Confirm there are no NAs
ncol(exonExpressionFem) #Count exons with data for all samples
```

```{r}
exonExpressionMale <- exonExpressionMod %>%
  filter(., grepl("M", sample_id, fixed = TRUE)) %>%
  select(where(~sum(!is.na(.)) > 0)) %>%
  column_to_rownames(., var = "sample_id") #Filter male samples only, remove exons with missing methylation data (NA) for any sample, and make sample_id 
head(exonExpressionMale) #Confirm formatting
sum(is.na(exonExpressionMale)) #Confirm there are no NAs
ncol(exonExpressionMale) #Count exons with data for all samples
```

## Exon methylation data

```{r}
exonMethylation <- read.delim("/home/shared/8TB_HDD_02/strigg/ceabigr/output/exon_summary_allsamples.txt", header = TRUE)
exonMethylation$sample <- gsub("_", "", exonMethylation$sample) #Remove _ from sample name
exonMethylationMod <- exonMethylation %>%
  unite(., exonID, 1:3) %>%
  select(., exonID, median_meth, sample) %>%
  distinct(.) %>%
  spread(., key = exonID, value = median_meth, fill = NA) #Create exonID column with chr-start-end, select only exonID, gene_cov, and sample_id information, filter unique rows, spread data for mixOmics formatting requirements
head(exonMethylationMod) #Confirm formatting
ncol(exonMethylationMod) #Count number of exons with data
```

### Sex-specific data

```{r}
exonMethylationFem <- exonMethylationMod %>%
  filter(., grepl("F", sample, fixed = TRUE)) %>%
  column_to_rownames(., var = "sample") #Filter female samples only and make sample rownames
exonMethylationFem <- as.data.frame(t(drop_na(as.data.frame(t(exonMethylationFem))))) #Remove columns with NAs
head(exonMethylationFem) #Confirm formatting
sum(is.na(exonMethylationFem)) #Confirm there are no NAs
ncol(exonMethylationFem) #Count exons with data for all samples
```

```{r}
exonMethylationMale <- exonMethylationMod %>%
  filter(., grepl("M", sample, fixed = TRUE)) %>%
  column_to_rownames(., var = "sample") #Filter male samples only and make sample rownames
exonMethylationMale <- as.data.frame(t(drop_na(as.data.frame(t(exonMethylationMale))))) #Remove columns with NAs
head(exonMethylationMale) #Confirm formatting
sum(is.na(exonMethylationMale)) #Confirm there are no NAs
ncol(exonMethylationMale) #Count exons with data for all samples
```

# Sparse PLS

Followed [`mixOmics` manual](https://mixomicsteam.github.io/Bookdown/pls.html#inputs-and-outputs-1) for analysis methods.

## Tune parameters

```{r}
exonResult.pls <- pls(exonMethylationFem, exonExpressionFem, ncomp = 4) #Run a PLS with a sufficient number of components
exonPerf.pls <- perf(exonResult.pls, validation = "Mfold", folds = 5,
                     progressBar = FALSE, nrepeat = 1000) #Use perf for repeated k-fold cross-validation to determine the number of components that should be included
plot(perf.pls$Q2.total) #Plot perf results
abline(h = 0.0975) #Only include components with value is â‰¤ 0.0975
```

## Run SPLS

```{r}
exonResult.spls <- spls(X = exonMethylationFem, Y = exonExpressionFem, 
                        keepX = c(25, 25), keepY = c(25,25)) #Run SPLS
```

```{r}
plotIndiv(exonResult.spls, group = metadata$Sex,
          legend = TRUE,
          legend.title = 'Sex',
          ind.names = metadata$Treatment,
          title = "Exon: sPLS") #Plot samples. Group by sex and add titles for treatment
```


```{r}
plotIndiv(exonResult.spls, group = metadata$Sex,
          rep.space = "XY-variate", legend = TRUE,
          legend.title = 'Sex',
          ind.names = metadata$Treatment,
          title = "Exon: sPLS") #Plot samples. Group by sex and add titles for treatment
```

```{r}
plotIndiv(exonResult.spls, group= metadata$Sex,
          pch = metadata$TreatmentN,
          rep.space = "XY-variate",  legend = TRUE,
          legend.title = "Sex", legend.title.pch = "Sex + Treatment",
          ind.names = FALSE,
          title = "Exon: sPLS") #Plot samples. Group by sex and add symbols for sex + treatment
```

```{r}
plotVar(exonResult.spls, cex = c(2,2)) #Plot variables
exonCoordinates <- plotVar(exonResult.spls, plot = FALSE) #Extract coordinates
```

```{r}
exonArrow <- plotArrow(exonResult.spls, group = metadata$TreatmentN, legend = TRUE,
                       X.label = 'PLS comp 1', Y.label = 'PLS comp 2') #Group by treatment and sex
exonArrow$data
plot(exonArrow)
```

## Component-level information

### Component 1

```{r}
exonSelectedVariables1 <- selectVar(exonResult.spls, comp = 1) #Selected variables for first component
exonSelectedVariables1$X$name #Selected exons from methylation data
exonSelectedVariables1$Y$name #Selected exons from expression data
```

```{r}
pdf("../output/exon-methylation-expression-mixomics/yrv-exonSPLS-cimPlot.pdf", width = 11, height = 8.5)
cim(exonResult.spls, comp = 1) #Clustered image map for first component
dev.off()
```

```{r}
plotLoadings(exonResult.spls, comp = 1, size.name = rel(0.5)) #Loadings for first component
```

### Component 2

```{r}
exonSelectedVariables2 <- selectVar(exonResult.spls, comp = 2) #Selected variables for second component
exonSelectedVariables2$X$name #Selected exons from methylation data
exonSelectedVariables2$Y$name #Selected exons from expression data
```

```{r}
cim(exonResult.spls, comp = 2) #Clustered image map for second component
```

```{r}
plotLoadings(exonResult.spls, comp = 2, size.name = rel(0.5)) #Loadings for second component
```

