---
title: "General-Methylation-Landscape"
output: html_document
---

# Prepare R Markdown document

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "../output/general-methylation-landscape") #Set root directory
```

# Obtain sample BEDgraphs

```{bash}
#Download 10x bedgraphs
wget -r \
--no-check-certificate --no-directories --no-parent --reject "index.html*" \
-P . \
-A "*10x.bedgraph" https://gannet.fish.washington.edu/seashell/bu-mox/scrubbed/120321-cvBS/
```

```{bash}
cd ../output/general-methylation-landscape
ls
```

```{bash}
cd ../output/general-methylation-landscape

for f in *10x.bedgraph
do
/home/shared/bedtools2/bin/sortBed \
-i ${f} \
> $(basename ${f%_10x.bedgraph})_10x.sort.bedgraph
done
```

```{bash}
ls *sort*
```

# Concatenate percent methylation information

## Correct for C->T SNPs

For each sample, I will use BS-Snper output to change the percent methylation for a C->T SNP to 0.

```{bash}
cd ../output/general-methylation-landscape

#Download 10x SNP
wget -r \
--no-check-certificate --no-directories --no-parent --reject "index.html*" \
-P . \
-A "*vcf" https://gannet.fish.washington.edu/seashell/bu-github/nb-2022/C_virginica/analyses/bsnp02/
```

## Create union BEDgraph

I will use `unionBedGraphs` to concatenate information for all loci across samples.

```{bash}
cd ../output/general-methylation-landscape

#Create union BEDgraph from sorted files
#Include a header
#Use N/A when there is no data for a CpG in a sample
#Define sample IDs
#Use sorted bedgraphs
#Save output
/home/shared/bedtools2/bin/unionBedGraphs \
-header \
-filler N/A \
-names 12M 13M 16F 19F 22F 23M 29F 31M 35F 36F 39F 3F 41F 44F 48M 50F 52F 53F 54F 59M 64M 6M 76F 77F 7M 9M \
-i \
*10x.sort.bedgraph \
> union_10x.bedgraph
```

```{bash}
cd ../output/general-methylation-landscape

#Check output
head union_10x.bedgraph
wc -l union_10x.bedgraph
```

