# code

`ceabigr/code`

---

- [02-te-all.sh](https://github.com/sr320/ceabigr/blob/main/code/02-te-all.sh): determining copy number of TE looking at WGBS coverage

- [07-TE-intersect-meth.Rmd](https://github.com/sr320/ceabigr/blob/main/code/07-TE-intersect-meth.Rmd): methylation level of TEs

- [10-TEm-TEcn.Rmd](https://github.com/sr320/ceabigr/blob/main/code/10-TEm-TEcn.Rmd): correlation of TE methylation and TE copy numnber

- [11-RNAseq-snps.Rmd](https://github.com/sr320/ceabigr/blob/main/code/11-RNAseq-snps.Rmd): finding SNPs in RNAseq data.

- [`18-differential-gene-expression-ballgown.Rmd`](https://github.com/sr320/ceabigr/blob/main/code/18-differential-gene-expression-ballgown.Rmd): Differential gene and transcript expression analysis using `ballgown` R package.

- [`19-exon-expression.Rmd`](https://github.com/sr320/ceabigr/blob/main/code/19-exon-expression.Rmd): Script to create a table of exon read counts per gene. Output files are in [`outputs/19-exon-expression`](https://github.com/sr320/ceabigr/tree/main/output/19-exon-expression).

- [`34-transcript-counts.Rmd`](https://github.com/sr320/ceabigr/blob/main/code/34-transcript-counts.Rmd): R Markdown file summarizing transcript counts on _per gene_ bases, across all samples treatments. Calculations include: median, mean, min, max, sum, and standard deviation.

- [`42-predominant_isoform-female_male.ipynb`](https://github.com/sr320/ceabigr/blob/main/code/42-predominant_isoform-female_male.ipynb): Jupyter notebook identifying differences in predominant isoforms between sexes.

- [`50-circos-formatting.Rmd`](https://github.com/sr320/ceabigr/blob/main/code/50-circos-formatting.Rmd): Creates tab-delimited Circos-formatted files for mean gene expression (FPKM) and mean methylation across samples/treatments.

- [60-expression-fpkm-coefficients_of_variation.Rmd](https://github.com/sr320/ceabigr/blob/main/code/60-expression-fpkm-coefficients_of_variation.Rmd): R Markdown file for calculating FPKM coefficients of variation (CoV) for genes and transcripts.