# code

`ceabigr/code`

---

## Manu-scripts

- [`20220131_cvir_hisat2-GCF_002022765.2_adult-oa-gonad.sh`](https://github.com/sr320/ceabigr/blob/main/code/20220131_cvir_hisat2-GCF_002022765.2_adult-oa-gonad.sh): Bash script to align RNA-seq data to _C.virginica_ genome using HISAT2.

- [`01.01-supplemental-tables.Rmd`](01.01-supplemental-tables.Rmd): R Markdown document for generating supplemental tables for manuscript. Output files are in [`supplemental-files/`](../supplemental-files/)

- [`18-differential-gene-expression-ballgown.Rmd`](https://github.com/sr320/ceabigr/blob/main/code/18-differential-gene-expression-ballgown.Rmd): Differential gene and transcript expression analysis using `ballgown` R package.

- [`42-predominant-isoform.Rmd)`](https://github.com/sr320/ceabigr/blob/main/code/42-predominant-isoform.Rmd): Transcriptional Noise Analysis

---

## Exploratory Scripts



- [02-te-all.sh](https://github.com/sr320/ceabigr/blob/main/code/02-te-all.sh): determining copy number of TE looking at WGBS coverage

- [07-TE-intersect-meth.Rmd](https://github.com/sr320/ceabigr/blob/main/code/07-TE-intersect-meth.Rmd): methylation level of TEs

- [10-TEm-TEcn.Rmd](https://github.com/sr320/ceabigr/blob/main/code/10-TEm-TEcn.Rmd): correlation of TE methylation and TE copy numnber

- [11-RNAseq-snps.Rmd](https://github.com/sr320/ceabigr/blob/main/code/11-RNAseq-snps.Rmd): finding SNPs in RNAseq data.

- [`17.1-GO-and-GOslim-CDS-Annotation.Rmd`](https://github.com/sr320/ceabigr/blob/main/code/17.1-GO-and-GOslim-CDS-Annotation.Rmd): Gene ontology (GO) and GOslim annotations of _C.virginica_ coding sequences (CDS).

- [`17-Swiss-Prot-Annotation.Rmd`](https://github.com/sr320/ceabigr/blob/main/code/17-Swiss-Prot-Annotation.Rmd): SwissProt annotation via NCBI BLASTp of _C.virginica_ protein and coding sequences (CDS).

- [`19-exon-expression.Rmd`](https://github.com/sr320/ceabigr/blob/main/code/19-exon-expression.Rmd): Script to create a table of exon read counts per gene. Output files are in [`outputs/19-exon-expression`](https://github.com/sr320/ceabigr/tree/main/output/19-exon-expression).

- [`34-transcript-counts.Rmd`](https://github.com/sr320/ceabigr/blob/main/code/34-transcript-counts.Rmd): R Markdown file summarizing transcript counts on _per gene_ bases, across all samples treatments. Calculations include: median, mean, min, max, sum, and standard deviation.

- [`42-predominant_isoform-female_male.ipynb`](https://github.com/sr320/ceabigr/blob/main/code/42-predominant_isoform-female_male.ipynb): Jupyter notebook identifying differences in predominant isoforms between sexes.

- [`50-circos-formatting.Rmd`](https://github.com/sr320/ceabigr/blob/main/code/50-circos-formatting.Rmd): Creates tab-delimited Circos-formatted files for mean gene expression (FPKM) and mean methylation across samples/treatments.

- [`52.1-rnaseq-relatedness.Rmd`](https://github.com/sr320/ceabigr/tree/main/code/52.1-rnaseq-relatedness.Rmd): Determines relatedness between RNAseq samples at the gene and transcript levels using gene/transcript counts generated by Stringtie.

- [`60-expression-fpkm-coefficients_of_variation.Rmd`](https://github.com/sr320/ceabigr/blob/main/code/60-expression-fpkm-coefficients_of_variation.Rmd): R Markdown file for calculating FPKM coefficients of variation (CoV) for genes and transcripts.

- [`65-exon-coverage.qmd`](https://github.com/sr320/ceabigr/blob/main/code/65-exon-coverage.qmd): Quarto Markdown file for determining natural log fold-change of exon expression for Exons 2 - 6, relative to Exon 1.
