# code

Code used in for each analysis can be found below.

## Genome Information and Feature Tracks



## Genetic Variation

- [`20220921-cvir-ceabigr-nextflow-epidiverse-snp.sh`](https://github.com/sr320/ceabigr/blob/main/code/20220921-cvir-ceabigr-nextflow-epidiverse-snp.sh): Bash script to run the Nextflow EpiDiverse/snp pipeline.

## Gene Activity

### Gene and Transcript Expression

- [`20220131_cvir_hisat2-GCF_002022765.2_adult-oa-gonad.sh`](https://github.com/sr320/ceabigr/blob/main/code/20220131_cvir_hisat2-GCF_002022765.2_adult-oa-gonad.sh): Bash script to align RNA-seq data to _C.virginica_ genome using HISAT2.
- [`18-differential-gene-expression-ballgown.Rmd`](https://github.com/sr320/ceabigr/blob/main/code/18-differential-gene-expression-ballgown.Rmd): Differential gene and transcript expression analysis using `ballgown` R package.

### Predominant Transcript Identification

- [`42-predominant-isoform.Rmd)`](https://github.com/sr320/ceabigr/blob/main/code/42-predominant-isoform.Rmd): Influence of Methylation on Predominant Transcript Shifts & Transcriptional Noise Analysis

### Alternative Splicing

## DNA Methylation Analysis

- [`bismark`](https://github.com/sr320/ceabigr/blob/main/code/01-bismark.sh): bismark code

- [`Genomic-Location-of-DML.Rmd`](https://github.com/sr320/ceabigr/blob/main/code/Genomic-Location-of-DML.Rmd):

- [`Genomic-Location-of-DML.ipynb`](https://github.com/sr320/ceabigr/blob/main/code/Genomic-Location-of-DML.ipynb):

### CpG locus-level Methylation Quantification

### Gene-level Methylation Quantification

## Methylation Influence on Gene Activity

## Influence of Methylation on Predominant Transcript Shifts

## Transcriptional Noise Analysis

- [`42-predominant-isoform.Rmd)`](https://github.com/sr320/ceabigr/blob/main/code/42-predominant-isoform.Rmd): Influence of Methylation on Predominant Transcript Shifts & Transcriptional Noise Analysis

## Influence of Methylation on Alternative Splicing


## Other scripts

- [`01.01-supplemental-tables.Rmd`](01.01-supplemental-tables.Rmd): R Markdown document for generating supplemental tables for manuscript. Output files are in [`supplemental-files/`](../supplemental-files/)
