# Data

`ceabigr/data`

---

- [`adult-meta.csv`](https://github.com/sr320/ceabigr/blob/main/data/adult-meta.csv): Originally from the [2018_L18-adult-methylation repo](https://github.com/epigeneticstoocean/2018_L18-adult-methylation) (might be private) organized by Alan Downey-Wall and Katie Lotterhos.

  Metadata for individual [_Crassostrea virginica_ (Eastern oyster)](https://en.wikipedia.org/wiki/Eastern_oyster) samples with the following columns:

    - `Sample.ID`: Unique sample ID for each individual designated in the following fashion:

      - `<integer>[FM]`:
      
        - `[FM]` indicates either `F`emale or `M`ale.

    - `OldSample.ID`: Unique old sample ID, which preceded `Sample.ID`. Same as `Sample.ID`, but each ID is preceded by `S`.

    - `Treatment`: OA treatment. Either `Control` or `Exposed`.

    - `Sex`: `F`emale or `M`ale.

    - `TreatmenN`: Numberical representation to identify combinations of `Sex` and `Treatment`:

      - `1`: Control, Male

      - `2`: Control, Female

      - `3`: Exposed, Male

      - `4`: Exposed, Female.

    - `Parent.ID`: Unique ID of individual parents with the following designations:

      - `[CE][FM]<integer>`:

        - `[CE]` indicates either `C`ontrol or `E`xposed.

        - `[FM]` indicates either `F`emale or `M`ale.

        - `<integer>` indicates a two digit integer.

- [`fpkm-all_transcripts.csv`](https://github.com/sr320/ceabigr/blob/main/data/fpkm-all_transcripts.csv): FPKM values for all transcripts from Ballgown expression analysis. See [`18-differential-gene-expression-ballgown.Rmd`](https://github.com/sr320/ceabigr/blob/main/code/18-differential-gene-expression-ballgown.Rmd).

- [`GCF_002022765.2_C_virginica-3.0-genes.blastx.outfmt6`](https://github.com/sr320/ceabigr/blob/main/data/GCF_002022765.2_C_virginica-3.0-genes.blastx.outfmt6): DIAMOND BLASTx of genes only ([`C_virginica-3.0_Gnomon_genes.bed`](https://github.com/sr320/ceabigr/blob/main/data/C_virginica-3.0_Gnomon_genes.bed)) against SwissProt database (downloaded 20230727). Top hit only for each query only.

- [`GCF_002022765.2_C_virginica-3.0_lncRNA.bed`](https://github.com/sr320/ceabigr/blob/632a54509603edd1157d43fae5e4e67212c7f034/data/GCF_002022765.2_C_virginica-3.0_lncRNA.bed): BED file with long, non-coding RNA (lncRNA) coordinates in the NCBI _C.virginica_ genome. Jupyter notebook (nbviewer):

  - [20220217-cvir-lncRNA_subsetting.ipynb](https://github.com/robertslab/code/blob/master/notebooks/sam/20220217-cvir-lncRNA_subsetting.ipynb)

- [`GCF_002022765.2_C_virginica-3.0_lncRNA.fa`](https://github.com/sr320/ceabigr/blob/632a54509603edd1157d43fae5e4e67212c7f034/data/GCF_002022765.2_C_virginica-3.0_lncRNA.fa): FastA file of all long, non-coding RNA (lncRNA) coordinates in the NCBI _C.virginica_ genome. Jupyter notebook (nbviewer):

  - [20220217-cvir-lncRNA_subsetting.ipynb](https://github.com/robertslab/code/blob/master/notebooks/sam/20220217-cvir-lncRNA_subsetting.ipynb)

- [`GCF_002022765.2_C_virginica-3.0_lncRNA.fa.fai`](https://github.com/sr320/ceabigr/blob/main/data/GCF_002022765.2_C_virginica-3.0_lncRNA.fa.fai): FastA index file for corresponding FastA with same prefix. Jupyter notebook (nbviewer):

  - [20220217-cvir-lncRNA_subsetting.ipynb](https://github.com/robertslab/code/blob/master/notebooks/sam/20220217-cvir-lncRNA_subsetting.ipynb)

- [`GCF_002022765.2_C_virginica-3.0_lncRNA.gff`](https://github.com/sr320/ceabigr/blob/632a54509603edd1157d43fae5e4e67212c7f034/data/GCF_002022765.2_C_virginica-3.0_lncRNA.gff): General Feature Format (GFF) file of all long, non-coding RNA (lncRNA) coordinates in the NCBI _C.virginica_ genome. Jupyter notebook (nbviewer):

  - [20220217-cvir-lncRNA_subsetting.ipynb](https://github.com/robertslab/code/blob/master/notebooks/sam/20220217-cvir-lncRNA_subsetting.ipynb)

- [`GCF_002022765.2_C_virginica-3.0_lncRNA.gtf`](https://github.com/sr320/ceabigr/blob/main/data/GCF_002022765.2_C_virginica-3.0_lncRNA.gtf): General Transfer Format (GTF) file of all long, non-coding RNA (lncRNA) coordinates in the NCBI _C.virginica_ genome. Jupyter notebook (nbviewer):

  - [20220217-cvir-lncRNA_subsetting.ipynb](https://github.com/robertslab/code/blob/master/notebooks/sam/20220217-cvir-lncRNA_subsetting.ipynb)

- [`gene_count_matrix.csv`](https://github.com/sr320/ceabigr/blob/main/data/gene_count_matrix.csv): Transcript count matrix generated by Stringtie's `prepDE.py3`. Suitable for import into DESeq2.

- [`transcript_count_matrix.csv`](https://github.com/sr320/ceabigr/blob/main/data/transcript_count_matrix.csv): Transcript count matrix generated by Stringtie's `prepDE.py3`. Suitable for import into DESeq2.

- [`whole_gx_table.csv`](https://github.com/sr320/ceabigr/blob/main/data/whole_gx_table.csv): Full table of gene expression, generated by Ballgwon.

- [`whole_tx_table.csv`](https://github.com/sr320/ceabigr/blob/main/data/whole_tx_table.csv): Full table of transcript expression, generated by Ballgown.